---
layout: default
title: Omics Integration and Systems Biology
---

#### Omics Integration and Systems Biology
9 - 12 September 2019, SciLifeLab, Tomtebodavagen 23A, 171 65 Stocholm

***Overview***

Course open for PhD students, postdocs, and researchers looking for an introduction to multi-omics integration and systems biology.

The aim of this workshop is to provide an integrated view of biological network construction and integration, constraint-based modelling, multi-omics integration and data-driven hypothesis generation.  A general description of different methods for analysing different omics data (e.g. transcriptomics and genomics) will be presented with some of the lectures discussing key methods in omics integration. The various techniques are discussed in terms of their rationale and applicability, with a particular focus on possible confounding factors. Many of the techniques presented in the workshop may be applied to different omics (e.g. transcriptomics, genomics, proteomics, metabolomics). Attendants will further get hands-on experience in their application. 

Some of the covered topics include:  

- Data management in omics studies;
- Condition-specific and personalized metabolic modeling based on integration of transcriptomic and proteomic data;  
- Identification of potential biomarkers and targetable genes;  
- Identification of key biological functions and pathways (e.g. gene set enrichment analysis);  
- Co-expression and association networks, gene modules, and network topology analysis;  
- Multi-omics integration, clustering and dimensionality reduction;  
- Application of key machine learning methods for multi-omics analysis. 

***Learning objectives***  
At the end of the course, students should:  
- Be able to integrate different omics and simulate biological functions using constraint-based models and [FBA](https://en.wikipedia.org/wiki/Flux_balance_analysis).
- Build biological networks based on different omics data, as well as integrated multi-omics networks, and perform basic topology analyses.
- Identify key methods for analysis and integration of omics data based on a given dataset.
- Be aware of important confounding factors and sources of bias.

***Materials***  
You can find all the [reading materials here][1].

***Course fee***  
A course fee of 2000SEK will be invoiced to accepted participants. This includes lunches, coffee, snacks, and course dinner. ***Please note that NBIS cannot invoice individuals.***

***Responsible teachers***  
- [Rui Benfeitas][8] (rui.benfeitas[at]scilifelab.se)
- [Nikolay Oskolkov](https://nbis.se/about/staff/nikolay-oskolkov) (nikolay.oskolkov[at]scilifelab.se)
- [Ashfaq Ali](https://nbis.se/about/staff/ashfaq-ali/) (ashfaq.ali[at]nbis.se)

***TAs***
- [Erik Fasterius][3]
- [Lokeshwaran Manoharan][4]
- [Roy Francis][5]
- [Mihail Anton][2]

Please contact [Rui Benfeitas][8] for questions about the course, or through [our Slack channel][7]

[1]: reading_materials.md
[2]: https://www.chalmers.se/en/staff/Pages/mihail-anton.aspx
[3]: https://nbis.se/about/staff/erik-fasterius/
[4]: https://nbis.se/about/staff/lokeshwaran-manoharan/
[5]: https://nbis.se/about/staff/roy-francis/
[6]: https://forms.gle/beDNyRoGib9jmFfz8
[7]: https://join.slack.com/t/omicsintegration/shared_invite/enQtNzQ1MDIzOTkxNTg2LThhMzlkZWY4MDgyM2NjOGVkNmU3YTgxMWE2MWIyNTMwODNmZjJlM2RlY2FhNDViMTU2YjQzMmE5YmUxOTY5ZmU
[8]: https://nbis.se/about/staff/rui-benfeitas