---
layout: default
title: Omics Integration and Systems Biology
---

#### <img border="0" src="https://s3-us-west-2.amazonaws.com/slack-files2/avatars/2019-09-12/751389607265_d59c0d58846bb2db7123_132.jpg" width="50" height="50"> Omics Integration and Systems Biology
[![DOI](https://zenodo.org/badge/172930292.svg)](https://zenodo.org/badge/latestdoi/172930292) <img border="0" src="https://www.svgrepo.com/show/305241/github.svg" width="20" height="20"> [Github repository](https://github.com/NBISweden/workshop_omics_integration)  
**Applications closed.** If you'd like to be notified of the next event please add your email [here](https://forms.gle/NhJNteYGCiZQSGwQ8)

<br>
<br>
<hr>

<img border="0" src="https://www.svgrepo.com/show/20800/event-date-and-time-symbol.svg" width="20" height="20"> 19 - 23 April 2021  
<img border="0" src="https://www.svgrepo.com/show/4199/placeholder-on-a-map.svg" width="20" height="20"> Online

The aim of this workshop is to provide an integrated view of biological network construction and integration, constraint-based modelling, multi-omics integration through Machine Learning, and data-driven hypothesis generation.  A general description of different methods for analysing different omics data (e.g. transcriptomics and genomics) will be presented with some of the lectures discussing key methods in their integration. The techniques will be discussed in terms of their rationale and applicability, with a particular focus on possible confounding factors. The course will also include hands-on sessions and invited speaker seminars.

**Audience** Course open for PhD students, postdocs, and researchers looking for an introduction to multi-omics integration and systems biology.

**Fee** This online training event has no fee. However, if you accept a position at the workshop and do not participate (no-show) you will be invoiced 2000 SEK. Please note that NBIS cannot invoice individuals.

<img border="0" src="https://www.svgrepo.com/show/26916/book.svg" width="20" height="20"> **Covered topics**  
- Condition-specific and personalized modeling through Genome-scale Metabolic models based on integration of transcriptomic, proteomic and metabolomic data;  
- Biological network inference, community and topology analysis and visualization
- Identification of key biological functions and pathways;  
- Identification of potential biomarkers and targetable genes through modeling and biological network analysis; 
- Application of key machine learning methods for multi-omics analysis including deep learning;  
- Multi-omics integration, clustering and dimensionality reduction;
- Similarity network fusion and Recommender systems;
- Integrated data visualization techniques

<img border="0" src="https://www.svgrepo.com/show/35663/tasks.svg" width="20" height="20"> **Learning objectives**  
At the end of the course, students should:  
- Be able to integrate different omics and simulate biological functions using constraint-based models and [FBA](https://en.wikipedia.org/wiki/Flux_balance_analysis).
- Build biological networks based on different omics data, as well as integrated multi-omics networks;
- Use different biological network analysis techniques to compare different cell-types or conditions (e.g. disease vs healthy);
- Identify key methods for analysis and integration of omics data based on a given dataset;
- Understand strengths and pitfalls of key machine learning techniques in multi-omic analysis;
- Understand confounding factors and sources of bias.

