---
layout: default
title:  Reading materials
---

#### Reading materials 

This page contains [recommended reading materials](#reading-materials) and details about some of 
the [software](#software) that will be used in the course.  
**! - recommended read before the course**

**Data management and data structures**
- [FAIR data guidelines][11]
- [Summarized Experiments][12] and [practical guide][28]
- [Multi Assay Experiments][13] and [practical guide][15]
- [Genomic Ranges][14]
  <br/><br/>

**Integration**
Note: if you cannot see one of the pages because you've reached the free member preview,
open the page in "Private Browsing" mode in your web browser.
- **!** [Feature selection][7] 
- **!** [Dimensionality reduction][8] 
- **!** [Supervised Omics integration][9] 
- **!** [Unsupervised Omics integration][10] 
- [Multi-omics approaches to disease][36]
- [Nikolay's notes from the board][42]
  <br/><br/>
  
**Biological Network Analysis and Network Topology**
- **!** [Network biology: understanding the cell's functional organization][34] 
- **!** [Network medicine: a network-based approach to human disease][39] 
- [Network biology concepts in complex disease comorbidities][41]
- **!**[Introduction to gene set enrichment analysis][6]
- [Deriving functions from gene sets][35] and functional analysis with [PIANO][18]
- [Gene co-expression analysis and conserved modules][38]
  <br/><br/>
  
**Genome-scale metabolic modeling**
- **!** [Overview of genome-scale metabolic models][40] 
- [A Systematic Evaluation of Methods for Tailoring Genome-Scale Metabolic Models][33]
- **!** Metabolic Models: [HMR and Metabolic Atlas][16] and [RECON3][17]  
- Tools for Metabolic modelling: [RAVEN][21], [COBRA][22], [openCOBRA][23] 
- Applications in Data integration: [pan cancer][19], [Obesity and diabetes][20] 
  <br/><br/>
  
**Data visualization**
- Circos plots in [command line with Circos][24], [R with circlize][25] or [DIABLO][26], and through [a web gui][27]  
- HivePlot [introduction][29] and [R tool with HiveR][30]  
- Cytoscape [introduction][31] **!** and communicate through R with [RCy3 Package][32]  
  <br/><br/>
  

#### Bonus

**Programming in R and Python**
If you would like to get started in one of the programming languages that we use, we regularly offer 
courses in [R and Python][1]. For instance:
- [Python programming with applications to bioinformatics][2] - Introductory Python course

You may also look at the following online materials:
- [RaukR 2019][3] - a comprehensive R course. New edition (Summer 2020) to be announced soon
- [Data Analysis and Visualization in Python][4] - an introductory course for Python
- [Intro to R and RStudio for Genomics][5]



[1]: https://nbis.se/training/events.html
[2]: https://www.scilifelab.se/events/python-programming-with-applications-to-bioinformatics-uppsala/
[3]: https://nbisweden.github.io/RaukR-2019/
[4]: https://datacarpentry.org/python-ecology-lesson/
[5]: https://datacarpentry.org/genomics-r-intro/
[6]: https://onlinelibrary.wiley.com/doi/full/10.1002/ajmg.b.32328
[7]: https://towardsdatascience.com/select-features-for-omics-integration-511390b7e7fd
[8]: https://towardsdatascience.com/reduce-dimensions-for-single-cell-4224778a2d67
[9]: https://towardsdatascience.com/supervised-omics-integration-2158e1a6d23f
[10]: https://medium.com/@nikolay.oskolkov/unsupervised-omics-integration-688bf8fa49bf
[11]: https://www.nature.com/articles/sdata201618
[12]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/
[13]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5679241/
[14]: https://bioconductor.org/packages/release/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html
[15]: http://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html
[16]: https://www.metabolicatlas.org
[17]: https://www.ncbi.nlm.nih.gov/pubmed/29457794
[18]: https://www.ncbi.nlm.nih.gov/pubmed/23444143
[19]: https://www.biorxiv.org/content/10.1101/050187v2
[20]: https://www.ncbi.nlm.nih.gov/pubmed/23630502
[21]: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6207324/
[22]: https://www.nature.com/articles/s41596-018-0098-2
[23]: https://opencobra.github.io/cobratoolbox/stable
[24]: http://circos.ca/documentation/course/
[25]: https://jokergoo.github.io/circlize_book/book/
[26]: http://www.biorxiv.org/content/biorxiv/early/2016/08/03/067611.full.pdf
[27]: http://mkweb.bcgsc.ca/tableviewer/visualize/
[28]: https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html
[29]: http://hiveplot.com
[30]: https://github.com/bryanhanson/HiveR
[31]: https://cytoscape.org/cytoscape-tutorials/protocols/tour-of-cytoscape/#/
[32]: https://bioconductor.org/packages/release/bioc/vignettes/RCy3/inst/doc/Overview-of-RCy3.html
[33]: https://www.cell.com/fulltext/S2405-4712(17)30010-8
[34]: https://www.nature.com/articles/nrg1272
[35]: stke.sciencemag.org/content/4/190/tr4?ijkey=W33OsCqwQHZ5k
[36]: https://link.springer.com/content/pdf/10.1186/s13059-017-1215-1.pdf
[37]: http://mural.maynoothuniversity.ie/1885/1/HamiltonMasonGraphTheory.pdf
[38]: https://science.sciencemag.org/content/302/5643/249.full
[39]: https://www.nature.com/articles/nrg2918
[40]: https://www.frontiersin.org/articles/10.3389/fphys.2015.00413/full
[41]: https://www.nature.com/articles/nrg.2016.87.pdf
[42]: https://github.com/NBISweden/workshop_omics_integration/blob/master/img/Nikolay_board.jpg
