<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ashfaq Ali" />


<title>Cytoscape_R</title>

<script src="Cytoscape_tutorial_files/header-attrs-2.10/header-attrs.js"></script>
<script src="Cytoscape_tutorial_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Cytoscape_tutorial_files/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="Cytoscape_tutorial_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Cytoscape_tutorial_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Cytoscape_tutorial_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Cytoscape_tutorial_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="Cytoscape_tutorial_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="Cytoscape_tutorial_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="Cytoscape_tutorial_files/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">Cytoscape_R</h1>
<h4 class="author">Ashfaq Ali</h4>
<h4 class="date">21/09/2020</h4>

</div>


<!-- rmd lab header -->
<!-- custom fonts -->
<p><link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;subset=latin-ext" rel="stylesheet"></p>
<p><br></p>
<p>This tutorial will cover R API for cytoscape through RCy3 package. People who are interested in Cytoscape GUI with can follow Lars Juhl Jensen’s labs tutorial(<a href="https://jensenlab.org/training/stringapp/" class="uri">https://jensenlab.org/training/stringapp/</a>).</p>
<div id="installation-and-testing-the-setup" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Installation and testing the setup</h1>
<p>If the packages required for this session are not available, please create the environment using the “renv.lock” file.¨</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>renv<span class="sc">::</span><span class="fu">restore</span>(<span class="at">lockfile =</span> <span class="st">&quot;renv.lock&quot;</span>)</span></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(dplyr))</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(RCy3))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(readxl))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(tidyverse))</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressMessages</span>(<span class="fu">library</span>(readxl))</span></code></pre></div>
</div>
<blockquote>
<p>Start the cytoscape app before running any of the follwoing code.</p>
</blockquote>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cytoscapePing</span>()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cytoscapeVersionInfo</span>()</span></code></pre></div>
<pre><code>##       apiVersion cytoscapeVersion 
##             &quot;v1&quot;          &quot;3.8.0&quot;</code></pre>
</div>
<p>Now close the cytoscape and install the string database app using R studio.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">installApp</span>(<span class="st">&quot;stringapp&quot;</span>)</span></code></pre></div>
</div>
<div id="installation-of-cytoscape-apps-using-r" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Installation of cytoscape apps using R</h2>
<p>To install cytoscape apps one can use the Cytoscape GUI through app manager or can install directly from R using the following code. The following step wont be necessary for this course.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>installation_responses <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co">#list of app to install</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>cyto_app_toinstall <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;clustermaker2&quot;</span>, <span class="st">&quot;enrichmentmap&quot;</span>, <span class="st">&quot;autoannotate&quot;</span>, <span class="st">&quot;wordcloud&quot;</span>, <span class="st">&quot;stringapp&quot;</span>, <span class="st">&quot;aMatReader&quot;</span>, <span class="st">&quot;yFiles Layout Algorithms&quot;</span>, <span class="st">&quot;enhancedGraphics&quot;</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>cytoscape_version <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>( <span class="fu">cytoscapeVersionInfo</span>()[<span class="st">&#39;cytoscapeVersion&#39;</span>],<span class="at">split =</span> <span class="st">&quot;</span><span class="sc">\\</span><span class="st">.&quot;</span>))</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(cytoscape_version) <span class="sc">==</span> <span class="dv">3</span> <span class="sc">&amp;&amp;</span> <span class="fu">as.numeric</span>(cytoscape_version[<span class="dv">1</span>]<span class="sc">&gt;=</span><span class="dv">3</span>) </span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>   <span class="sc">&amp;&amp;</span> <span class="fu">as.numeric</span>(cytoscape_version[<span class="dv">2</span>]<span class="sc">&gt;=</span><span class="dv">7</span>)){</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(cyto_app_toinstall)){</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>    <span class="co">#check to see if the app is installed.  Only install it if it hasn&#39;t been installed</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="sc">!</span><span class="fu">grep</span>(<span class="fu">commandsGET</span>(<span class="fu">paste</span>(<span class="st">&quot;apps status app=</span><span class="sc">\&quot;</span><span class="st">&quot;</span>, cyto_app_toinstall[<span class="dv">1</span>],<span class="st">&quot;</span><span class="sc">\&quot;</span><span class="st">&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)), </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>             <span class="at">pattern =</span> <span class="st">&quot;status: Installed&quot;</span>)){</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>      installation_response <span class="ot">&lt;-</span><span class="fu">commandsGET</span>(<span class="fu">paste</span>(<span class="st">&quot;apps install app=</span><span class="sc">\&quot;</span><span class="st">&quot;</span>, </span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>                                                cyto_app_toinstall[i],<span class="st">&quot;</span><span class="sc">\&quot;</span><span class="st">&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>))</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>      installation_responses <span class="ot">&lt;-</span> <span class="fu">c</span>(installation_responses,installation_response)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span>{</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>      installation_responses <span class="ot">&lt;-</span> <span class="fu">c</span>(installation_responses,<span class="st">&quot;already installed&quot;</span>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  installation_summary <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">name =</span> cyto_app_toinstall, </span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">status =</span> installation_responses)</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">list</span>(installation_summary),</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">booktabs =</span> <span class="cn">TRUE</span>, <span class="at">caption =</span> <span class="st">&#39;A Summary of automated app installation&#39;</span></span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
</div>
<div id="intuition-behind-using-rcy3-to-analyze-networks-with-cytoscape." class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Intuition behind using RCy3 to analyze networks with cytoscape.</h2>
<p>Cytoscape(www.cytoscape.org) is probably the most popular applications for network analysis and visualization with GUI. In this session, we will learn about new capabilities to integrate Cytoscape into programmatic workflows and pipelines using R. Apart from programmatic workflow, the access through R allows direct access to network analyses as being part of an ongoing analyses instead of a separate instance. This also makes analyses steps more explicit hence more reproducible. There are community provided scripts available for various kinds of analyses at (<a href="https://github.com/cytoscape/cytoscape-automation/tree/master/for-scripters/R/notebooks" class="uri">https://github.com/cytoscape/cytoscape-automation/tree/master/for-scripters/R/notebooks</a>). If you are interested in the type of analyses that suits your project more, you may find some scripts related to that here. This tutorial is just a quick hands on introduction to using Cytoscape using from R. Cytoscape has amazing amount of functionality and only a fraction of it can be covered here.</p>
<p>We have already been introduced to network biology themes and concepts, therefor we will translate these into Cytoscape terms for practical applications. This tutorial will be a hands-on session working towards accessing and controlling Cytoscape from R to perform a network analysis. We will go through the following themes that are usually required in network visualization and analyses.</p>
<ul>
<li>retrieve networks for Genes/ proteins of interest</li>
<li>layout and visually style the resulting networks</li>
<li>import external data and map them onto a network</li>
<li>filter networks and customize styling</li>
<li>perform co-expression analyses and create your own co-expression network</li>
<li>visualize the results and map attributes to user created networks</li>
</ul>
</div>
</div>
<div id="data-preparation" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Data preparation</h1>
<p>We will be using data from the Charge consortium in which they have analyzed quite large cohort with association of methylation and gene expression with Age. We will use some of the results for visualization in Cytoscape. In addition there is a list of proteins and coefficients from a proteomics data-set provided in the <strong>/Data/AgeingTopProteins.txt</strong>. All these data aim to identify markers associated with age. We can try and find overlaps in these data-sets using network analyses in cytoscape.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#setwd(&quot;/Users/cob-aaf/Documents/GitHub/session_visualization/&quot;)</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth <span class="ot">&lt;-</span> <span class="fu">read_excel</span>(<span class="st">&quot;./Data/AgeExpMethSig.xlsx&quot;</span>,   <span class="at">skip =</span> <span class="dv">0</span>) <span class="do">## Provide the path to the file with data</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>Age_prot <span class="ot">&lt;-</span> <span class="fu">read_table2</span>(<span class="st">&quot;./Data/AgeingTopProteins.txt&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>Age_exp <span class="ot">&lt;-</span> AgeingExpMeth <span class="sc">%&gt;%</span>  <span class="fu">arrange</span>(<span class="st">`</span><span class="at">ExpAgeP-value</span><span class="st">`</span>) <span class="sc">%&gt;%</span> <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">100</span>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>Age_meth <span class="ot">&lt;-</span> AgeingExpMeth <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(<span class="st">`</span><span class="at">MethAgeP-value</span><span class="st">`</span>) <span class="sc">%&gt;%</span> <span class="fu">slice_head</span>(<span class="at">n=</span><span class="dv">100</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Age_exp) <span class="ot">&lt;-</span> Age_exp<span class="sc">$</span>Gene</span></code></pre></div>
</div>
</div>
<div id="get-help" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Get help</h1>
<p>To get an overview of what commands and functions are available through “RCy3”, please have a look at the help pages. More details with examples are also available through package vignettes.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">help</span>(<span class="at">package=</span>RCy3)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">browseVignettes</span>(<span class="st">&quot;RCy3&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="retrieve-networks-for-genesproteins-of-interest" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Retrieve networks for genes/proteins of interest</h1>
<p>To retrieve a network from the StringDB based on your own genes/proteins of interest from your own data, you need to construct a query command as following.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>AgeingExpe_string_interaction_cmd <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;string protein query taxonID=9606 cutoff=0.9 query=&#39;</span>, <span class="fu">paste</span>(Age_exp<span class="sc">$</span>Gene, <span class="at">collapse=</span><span class="st">&quot;,&quot;</span>),<span class="st">&#39;&quot;&#39;</span>,<span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">commandsGET</span>(AgeingExpe_string_interaction_cmd )</span></code></pre></div>
<pre><code>## [1] &quot;Loaded network &#39;STRING network&#39; with 109 nodes and 90 edges&quot;</code></pre>
</div>
<p>To get a better understanding of how StringDB can be constructed, read the <code>commandsGET</code> help pages.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>?commandsGET</span></code></pre></div>
</div>
<p>One can set a name to the retrieved network, as cytoscape gives your network a generic name. And if you have retrieved several networks, it may get confusing after a while.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">renameNetwork</span>(<span class="st">&quot;AgeingExpeNet&quot;</span>)</span></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>AgeingMeth_string_interaction_cmd <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;string protein query taxonID=9606 cutoff=0.9 query=&#39;</span>, <span class="fu">paste</span>(Age_meth<span class="sc">$</span>Gene, <span class="at">collapse=</span><span class="st">&quot;,&quot;</span>),<span class="st">&#39;&quot;&#39;</span>,<span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">commandsGET</span>(AgeingMeth_string_interaction_cmd)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">renameNetwork</span>(<span class="st">&quot;AgeingMethNet&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Loaded network &#39;STRING network - 1&#39; with 109 nodes and 97 edges&quot;</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>AgeingProt_string_interaction_cmd <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&#39;string protein query taxonID=9606 limit=150 cutoff=0.9 query=&#39;</span>, <span class="fu">paste</span>(Age_prot<span class="sc">$</span>protein, <span class="at">collapse=</span><span class="st">&quot;,&quot;</span>),<span class="st">&#39;&quot;&#39;</span>,<span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>test <span class="ot">&lt;-</span><span class="fu">commandsGET</span>(AgeingProt_string_interaction_cmd)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">renameNetwork</span>(<span class="st">&quot;AgeingProtNet&quot;</span>)</span></code></pre></div>
</div>
<p>To get an overview of what networks are present in cytoscape workspace use following:</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getNetworkList</span>()</span></code></pre></div>
<pre><code>## [1] &quot;AgeingMethNet&quot; &quot;AgeingExpeNet&quot; &quot;AgeingProtNet&quot;</code></pre>
</div>
<p>To work with a respective network, we need to set the environment we wish to work with.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setCurrentNetwork</span>(<span class="at">network=</span><span class="st">&quot;AgeingExpeNet&quot;</span>)</span></code></pre></div>
</div>
<p>Save the original network to a file.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">file.path</span>( <span class="st">&quot;Data&quot;</span>,<span class="st">&quot;ExpnetworkOriginal.png&quot;</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co"># initial_string_network_png_file_name &lt;- file.path(getwd(), &quot;/Path/to/your/network.png&quot;)</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">exportImage</span>(path, <span class="at">type =</span> <span class="st">&quot;png&quot;</span>, <span class="at">resolution =</span> <span class="dv">300</span>)</span></code></pre></div>
</div>
<p><strong><em>Cytoscape will hang waiting for user response if the network file already exists. You have to remove it first</em></strong></p>
</div>
<div id="do-on-your-own" class="section level1" number="5">
<h1><span class="header-section-number">5</span> Do on your own</h1>
<ol style="list-style-type: decimal">
<li>Can you save the <strong>“AgeingMethNet”</strong> and <strong>“AgeingProtNet”</strong> to the respective files?</li>
<li>Use your desired node attributes among the cytoscape network columns to improve visualization.</li>
</ol>
</div>
<div id="change-network-layout-and-visually-style-the-networks" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Change network layout and visually style the networks</h1>
<p>Now that we have a few networks loaded in cytoscape, let us work with those based on our interest e.g., change the layout of the network and save to a new file.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>layouts <span class="ot">&lt;-</span> <span class="fu">getLayoutNames</span>() <span class="do">## Access the available lyouts in cytoscape</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">layoutNetwork</span>(layouts[<span class="dv">4</span>]) <span class="do">## Change the layout to circular</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#exportImage(path, type = &quot;png&quot;, resolution = 300) ## Cytoscape will hang as you are trying to overwite an existing file</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>path <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(), <span class="st">&quot;/Data/&quot;</span>,<span class="st">&quot;ExpnetworkOriginalCircular.png&quot;</span>) <span class="co"># update the path</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">exportImage</span>(<span class="st">&quot;./Data/ExpnetworkOriginalCircular1.png&quot;</span>, <span class="at">type =</span> <span class="st">&quot;png&quot;</span>, <span class="at">resolution =</span> <span class="dv">300</span>) <span class="do">## Cytoscape will hang as you are trying to overwite an existing file</span></span></code></pre></div>
<pre><code>##                                                                                                              file 
## &quot;/Users/cob-aaf/Documents/GitHub/omicsintELexir2021/session_visualization/./Data/ExpnetworkOriginalCircular1.png&quot;</code></pre>
</div>
<p>Get available parameters for a specific layout.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getLayoutPropertyNames</span>(<span class="at">layout.name=</span><span class="st">&#39;force-directed&#39;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Available arguments for &#39;layout force-directed&#39;:&quot;
##  [1] &quot;defaultEdgeWeight&quot;        &quot;defaultNodeMass&quot;         
##  [3] &quot;defaultSpringCoefficient&quot; &quot;defaultSpringLength&quot;     
##  [5] &quot;edgeAttribute&quot;            &quot;isDeterministic&quot;         
##  [7] &quot;maxWeightCutoff&quot;          &quot;minWeightCutoff&quot;         
##  [9] &quot;network&quot;                  &quot;nodeAttribute&quot;           
## [11] &quot;nodeList&quot;                 &quot;numIterations&quot;           
## [13] &quot;singlePartition&quot;          &quot;type&quot;</code></pre>
</div>
<p>Apply the layout by changing parameters again</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">layoutNetwork</span>(<span class="st">&#39;force-directed defaultSpringCoefficient=0.0000008 defaultSpringLength=70&#39;</span>)</span></code></pre></div>
</div>
<p>Overlay our expression data on the String network.
To do this we will be using the loadTableData function from RCy3. It is important to make sure that that your identifiers types match up. You can check what is used by String by pulling in the column names of the node attribute table.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getTableColumnNames</span>(<span class="st">&#39;node&#39;</span>)</span></code></pre></div>
<pre><code>##  [1] &quot;SUID&quot;                                &quot;shared name&quot;                        
##  [3] &quot;name&quot;                                &quot;selected&quot;                           
##  [5] &quot;stringdb::canonical name&quot;            &quot;display name&quot;                       
##  [7] &quot;stringdb::full name&quot;                 &quot;stringdb::database identifier&quot;      
##  [9] &quot;stringdb::description&quot;               &quot;@id&quot;                                
## [11] &quot;stringdb::namespace&quot;                 &quot;stringdb::node type&quot;                
## [13] &quot;query term&quot;                          &quot;stringdb::sequence&quot;                 
## [15] &quot;stringdb::species&quot;                   &quot;stringdb::STRING style&quot;             
## [17] &quot;stringdb::enhancedLabel Passthrough&quot; &quot;compartment::cytoskeleton&quot;          
## [19] &quot;compartment::cytosol&quot;                &quot;compartment::endoplasmic reticulum&quot; 
## [21] &quot;compartment::endosome&quot;               &quot;compartment::extracellular&quot;         
## [23] &quot;compartment::golgi apparatus&quot;        &quot;compartment::lysosome&quot;              
## [25] &quot;compartment::mitochondrion&quot;          &quot;compartment::nucleus&quot;               
## [27] &quot;compartment::peroxisome&quot;             &quot;compartment::plasma membrane&quot;       
## [29] &quot;stringdb::interactor score&quot;          &quot;stringdb::structures&quot;               
## [31] &quot;target::development level&quot;           &quot;target::family&quot;                     
## [33] &quot;tissue::adrenal gland&quot;               &quot;tissue::blood&quot;                      
## [35] &quot;tissue::bone&quot;                        &quot;tissue::bone marrow&quot;                
## [37] &quot;tissue::eye&quot;                         &quot;tissue::gall bladder&quot;               
## [39] &quot;tissue::heart&quot;                       &quot;tissue::intestine&quot;                  
## [41] &quot;tissue::kidney&quot;                      &quot;tissue::liver&quot;                      
## [43] &quot;tissue::lung&quot;                        &quot;tissue::muscle&quot;                     
## [45] &quot;tissue::nervous system&quot;              &quot;tissue::pancreas&quot;                   
## [47] &quot;tissue::saliva&quot;                      &quot;tissue::skin&quot;                       
## [49] &quot;tissue::spleen&quot;                      &quot;tissue::stomach&quot;                    
## [51] &quot;tissue::thyroid gland&quot;               &quot;tissue::urine&quot;</code></pre>
</div>
<p>The column “display name” contains gene names which are also found in our Ageing Expression data-set.
To import our expression data we will match our data-set to the “display name” node attribute.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setCurrentNetwork</span>(<span class="at">network=</span><span class="st">&quot;AgeingExpeNet&quot;</span>)</span></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>?loadTableData</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">loadTableData</span>(<span class="at">data=</span> <span class="fu">as.data.frame</span>(Age_exp),  <span class="at">table.key.column =</span> <span class="st">&quot;display name&quot;</span>, <span class="at">data.key.column=</span> <span class="st">&quot;Gene&quot;</span>)  <span class="co">#default data.frame key is row.names</span></span></code></pre></div>
<pre><code>## [1] &quot;Success: Data loaded in defaultnode table&quot;</code></pre>
</div>
<p>Modify the Visual Style Create your own visual style to visualize your expression data on the String network.</p>
<p>Start with a default style</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>my.style.name <span class="ot">=</span> <span class="st">&quot;AgeingStyle&quot;</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>defaults.list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">NODE_SHAPE=</span><span class="st">&quot;ellipse&quot;</span>,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">NODE_SIZE=</span><span class="dv">60</span>,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">NODE_FILL_COLOR=</span><span class="st">&quot;#AAAAAA&quot;</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">EDGE_TRANSPARENCY=</span><span class="dv">120</span>)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>node.label.map <span class="ot">&lt;-</span> <span class="fu">mapVisualProperty</span>(<span class="st">&#39;node label&#39;</span>,<span class="st">&#39;display name&#39;</span>,<span class="st">&#39;p&#39;</span>) <span class="co"># p for passthrough; nothing else needed</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a><span class="fu">createVisualStyle</span>(my.style.name, defaults.list, <span class="fu">list</span>(node.label.map))</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="fu">setVisualStyle</span>(<span class="at">style.name=</span>my.style.name)</span></code></pre></div>
<pre><code>##                 message 
## &quot;Visual Style applied.&quot;</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>min.Exp.zscore <span class="ot">=</span> <span class="fu">min</span>(Age_exp<span class="sc">$</span><span class="st">`</span><span class="at">Z-score</span><span class="st">`</span>, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>max.Exp.zscore  <span class="ot">=</span> <span class="fu">max</span>(Age_exp<span class="sc">$</span><span class="st">`</span><span class="at">Z-score</span><span class="st">`</span>, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co">#mid..Exp.zscore = mean</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>data.values <span class="ot">=</span> <span class="fu">c</span>(min.Exp.zscore, <span class="fu">mean</span>(min.Exp.zscore,max.Exp.zscore) ,max.Exp.zscore)</span></code></pre></div>
</div>
<p>Next, we use the RColorBrewer package to help us pick good colors to pair with our data values.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">display.brewer.all</span>(<span class="fu">length</span>(data.values), <span class="at">colorblindFriendly=</span><span class="cn">TRUE</span>, <span class="at">type=</span><span class="st">&quot;div&quot;</span>) <span class="co"># div,qual,seq,all</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>node.colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="fu">length</span>(data.values), <span class="st">&quot;RdYlBu&quot;</span>)))</span></code></pre></div>
<img src="Cytoscape_tutorial_files/figure-html/external_data_color-1.svg" width="672" style="display: block; margin: auto auto auto 0;" />
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setNodeColorMapping</span>(<span class="st">&quot;Z-score&quot;</span>, data.values, node.colors, <span class="at">style.name=</span>my.style.name)</span></code></pre></div>
<pre><code>## NULL</code></pre>
</div>
<p>Change the size of the node to be correlated with the Age p-value.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setNodeSizeMapping</span>(<span class="at">table.column =</span> <span class="st">&#39;ExpAgeP-value&#39;</span>, </span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">table.column.values =</span> <span class="fu">c</span>(<span class="fu">min</span>(Age_exp<span class="sc">$</span><span class="st">`</span><span class="at">ExpAgeP-value</span><span class="st">`</span>), </span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="fu">mean</span>(Age_exp<span class="sc">$</span><span class="st">`</span><span class="at">ExpAgeP-value</span><span class="st">`</span>), </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>                                           <span class="fu">max</span>(Age_exp<span class="sc">$</span><span class="st">`</span><span class="at">ExpAgeP-value</span><span class="st">`</span>)), </span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">sizes =</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">60</span>, <span class="dv">150</span>), <span class="at">mapping.type =</span> <span class="st">&quot;c&quot;</span>, <span class="at">style.name =</span> my.style.name)</span></code></pre></div>
<pre><code>## NULL</code></pre>
</div>
<div id="creating-a-degree-filter" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Creating a degree filter</h2>
<p>Every node in a network has a Degree property, which corresponds to the number of edges connecting the node to other nodes, either as a target or source. Filtering based on node degree is a useful way to remove nodes with too few (or too many) connections.</p>
<p>In this example we want to exclude low degree nodes, e.g., those with only 0, 1 or 2 connections:</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">createDegreeFilter</span>(<span class="st">&#39;degree filter&#39;</span>, <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">2</span>), <span class="st">&#39;IS_NOT_BETWEEN&#39;</span>)</span></code></pre></div>
<pre><code>## $nodes
##  [1] &quot;9606.ENSP00000398124&quot; &quot;9606.ENSP00000355759&quot; &quot;9606.ENSP00000221801&quot;
##  [4] &quot;9606.ENSP00000278616&quot; &quot;9606.ENSP00000265165&quot; &quot;9606.ENSP00000479618&quot;
##  [7] &quot;9606.ENSP00000368880&quot; &quot;9606.ENSP00000358563&quot; &quot;9606.ENSP00000225916&quot;
## [10] &quot;9606.ENSP00000376436&quot; &quot;9606.ENSP00000263253&quot; &quot;9606.ENSP00000262367&quot;
## [13] &quot;9606.ENSP00000362649&quot; &quot;9606.ENSP00000335153&quot; &quot;9606.ENSP00000332973&quot;
## [16] &quot;9606.ENSP00000269305&quot; &quot;9606.ENSP00000341551&quot; &quot;9606.ENSP00000266970&quot;
## [19] &quot;9606.ENSP00000264279&quot; &quot;9606.ENSP00000344456&quot;
## 
## $edges
## [1] NA</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">createSubnetwork</span>(<span class="at">subnetwork.name =</span><span class="st">&#39;Ageing: highly connected nodes&#39;</span>)</span></code></pre></div>
<pre><code>## network 
##   14927</code></pre>
</div>
</div>
<div id="creating-a-column-filter" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Creating a column filter</h2>
<p>We could also filter the network based on any of the column data available. Let us filter nodes that come from “target” database belonging to “family” “TF”.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">createColumnFilter</span>(<span class="at">filter.name=</span><span class="st">&#39;transcription factors&#39;</span>, <span class="at">column=</span><span class="st">&quot;target::family&quot;</span>, <span class="st">&quot;TF&quot;</span>, <span class="st">&quot;IS&quot;</span>, <span class="at">network =</span> <span class="st">&quot;Ageing: highly connected nodes&quot;</span>)</span></code></pre></div>
<pre><code>## $nodes
## [1] &quot;9606.ENSP00000479618&quot; &quot;9606.ENSP00000368880&quot; &quot;9606.ENSP00000398124&quot;
## [4] &quot;9606.ENSP00000332973&quot; &quot;9606.ENSP00000376436&quot; &quot;9606.ENSP00000269305&quot;
## [7] &quot;9606.ENSP00000265165&quot; &quot;9606.ENSP00000341551&quot;
## 
## $edges
## [1] NA</code></pre>
</div>
<p>Now we will select the nodes from the family and color them differently in the network.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">getSelectedNodes</span>()</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">setNodeColorBypass</span>(nodes,<span class="at">new.colors=</span><span class="st">&#39;#FF0088&#39;</span>)</span></code></pre></div>
</div>
<p>Here the string matching is based on discrete values but it can be from a continuous variable?</p>
<p><strong>Can you select a few nodes based on your favorite column and see whether you can map some column as node attributes?</strong></p>
</div>
<div id="generate-own-networks-using-co-expression-analyse" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Generate own networks using co-expression analyse</h2>
<p>With the following block of code one can perform co-expression analyses. Here you are already provided with correlation matrix that you can use to follow the steps.</p>
<p><strong>This part will not work if you try to copy and paste in the R console, you have to provide the expression matrix of your own data</strong></p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co">#scale the data if required</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>RNASeq_expression <span class="ot">&lt;-</span> <span class="fu">scale</span>(NASeq_expression, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co">#rownames(RNASeq_expression) &lt;- RNASeq_expression_matrix$Name</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>RNAseq_correlation_matrix <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="fu">t</span>(RNASeq_expression), <span class="at">method=</span><span class="st">&quot;pearson&quot;</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="co">#set the diagonal of matrix to zero - eliminate self-correlation</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>RNAseq_correlation_matrix[ <span class="fu">row</span>(RNAseq_correlation_matrix) <span class="sc">==</span> <span class="fu">col</span>(RNAseq_correlation_matrix) ] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a><span class="co"># set all correlations that are less than 0.7 to zero</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>RNAseq_correlation_matrix[<span class="fu">which</span>(RNAseq_correlation_matrix<span class="sc">&lt;</span><span class="fl">0.60</span>)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a><span class="co">#get rid of rows and columns that have no correlations with the above thresholds</span></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>RNAseq_correlation_matrix <span class="ot">&lt;-</span> RNAseq_correlation_matrix[<span class="fu">which</span>(<span class="fu">rowSums</span>(RNAseq_correlation_matrix) <span class="sc">!=</span> <span class="dv">0</span>),</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>                          <span class="fu">which</span>(<span class="fu">colSums</span>(RNAseq_correlation_matrix) <span class="sc">!=</span><span class="dv">0</span>)]</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a><span class="co">#correlation_filename &lt;- file.path(getwd(),  &quot;Data&quot;, &quot;Protein_correlation_matrix.txt&quot;)</span></span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>( RNAseq_correlation_matrix ,  <span class="at">file =</span> correlation_filename, <span class="at">col.names  =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">&quot;</span><span class="sc">\t</span><span class="st">&quot;</span>, <span class="at">quote=</span><span class="cn">FALSE</span>)</span></code></pre></div>
</div>
<p>Use the CyRest call to access the aMatReader functionality.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>correlation_filename <span class="ot">&lt;-</span> <span class="fu">file.path</span>(<span class="fu">getwd</span>(),  <span class="st">&quot;Data&quot;</span>, <span class="st">&quot;Protein_correlation_matrix.txt&quot;</span>) </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>amat_url <span class="ot">&lt;-</span> <span class="st">&quot;aMatReader/v1/import&quot;</span></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>amat_params <span class="ot">=</span> <span class="fu">list</span>(<span class="at">files =</span> <span class="fu">list</span>(correlation_filename),</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">delimiter =</span> <span class="st">&quot;TAB&quot;</span>,</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">undirected =</span> <span class="cn">FALSE</span>,</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">ignoreZeros =</span> <span class="cn">TRUE</span>,</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">interactionName =</span> <span class="st">&quot;correlated with&quot;</span>,</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">rowNames =</span> <span class="cn">FALSE</span></span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>                   )</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>response <span class="ot">&lt;-</span> <span class="fu">cyrestPOST</span>(<span class="at">operation =</span> amat_url, <span class="at">body =</span> amat_params, <span class="at">base.url =</span> <span class="st">&quot;http://localhost:1234&quot;</span>)</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>current_network_id <span class="ot">&lt;-</span> response<span class="sc">$</span>data[<span class="st">&quot;suid&quot;</span>]</span></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">layoutNetwork</span>(<span class="st">&#39;cose&#39;</span>,<span class="at">network =</span> <span class="fu">as.numeric</span>(current_network_id))</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">renameNetwork</span>(<span class="at">title =</span> <span class="st">&quot;Coexpression_network&quot;</span>,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">network =</span> current_network_id)</span></code></pre></div>
</div>
<p>One can map column attributes as we did in case of string database based on our desired attributes.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co">#?loadTableData</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">loadTableData</span>(<span class="at">data=</span> <span class="fu">as.data.frame</span>(Age_prot),  <span class="at">table.key.column =</span> <span class="st">&quot;name&quot;</span>, <span class="at">data.key.column=</span> <span class="st">&quot;protein&quot;</span>)  <span class="co">#default data.frame key is row.names</span></span></code></pre></div>
<pre><code>## [1] &quot;Success: Data loaded in defaultnode table&quot;</code></pre>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">createColumnFilter</span>(<span class="at">filter.name=</span><span class="st">&#39;Fold change&#39;</span>, <span class="at">column=</span><span class="st">&quot;logFC&quot;</span>, <span class="fl">0.01</span>, <span class="st">&quot;GREATER_THAN&quot;</span>, <span class="at">network =</span> <span class="st">&quot;Coexpression_network&quot;</span>)</span></code></pre></div>
</div>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">getSelectedNodes</span>()</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="fu">setNodeColorBypass</span>(nodes,<span class="at">new.colors=</span><span class="st">&#39;#FF0088&#39;</span>)</span></code></pre></div>
</div>
</div>
<div id="bonus-exercises" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Bonus exercises</h2>
<p>Create consensus scores to the respective gene and methylation data-sets, create and network based on string db. Map attributes where both gene expression and methylation data agrees most.</p>
<div class="block-title-parent">
<div class="block-title small">
R
</div>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create attributes for the &quot;Age&quot; dataset</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>ExpAgeUp <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>ExpAgeDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[+]&quot;</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>ExpAgeDown <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>ExpAgeDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[-]&quot;</span>)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>MethAgeUp <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>MethAgeDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[+]&quot;</span>)</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>MethAgeDown <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>MethAgeDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[-]&quot;</span>)</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>ExpMethUp <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>ExpMethDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[+]&quot;</span>)</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>AgeingExpMeth<span class="sc">$</span>ExpMethDown <span class="ot">&lt;-</span> AgeingExpMeth<span class="sc">$</span>ExpMethDirection <span class="sc">%&gt;%</span> stringi<span class="sc">::</span><span class="fu">stri_count</span>( <span class="at">regex =</span> <span class="st">&quot;[-]&quot;</span>)</span></code></pre></div>
</div>
<hr />
</div>
<div id="end-the-session" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> End the session</h2>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
