---
title: "Tutorial"
output: learnr::tutorial
runtime: shiny_prerendered
---


```{r}
# suppressMessages(library(dplyr))
# suppressMessages(library(purrr))
suppressMessages(library(stringi))
suppressMessages(library(tidyr))
suppressMessages(library(tidyverse))
suppressMessages(library(plyranges))
suppressMessages(library(circlize))
suppressMessages(library(IRanges))
suppressMessages(library(data.table))

# library(HiveR)
# library(rgl)
```

# Data visualization


Data visualization is part art and part science. The challenge is to get the art right without getting the science wrong and vice versa.

Anscombe’s quartet notwithstanding, and especially for large volumes of data, summary statistics and model estimates should be thought of as tools that we use to deliberately simplify things in a way that lets us see past a cloud of data points shown in a figure. We will not automatically get the right answer to our questions just by looking.

Tufte’s message is sometimes frustrating, but it is consistent:

Graphical excellence is the well-designed presentation of interesting data—a matter of substance, of statistics, and of design … [It] consists of complex ideas communicated with clarity, precision, and efficiency. … [It] is that which gives to the viewer the greatest number of ideas in the shortest time with the least ink in the smallest space … [It] is nearly always multivariate … And graphical excellence requires telling the truth about the data. (Tufte, 1983, p. 51).

To get more insight into data visualization and principles one should follow, please refer to following two books from where I have copied the above text.

1. Data Visualization: A practical introduction by Kieran Healy (https://socviz.co/lookatdata.html)
2. Fundamentals of Data Visualization by Claus O. Wilke (https://clauswilke.com/dataviz/)


# Circos Plots

The circos plot capability to plot genomic data with chromosome coordinates that requires chromosom *start* and the *end* site indicating the length of the chromosome. Each chromosom will occupy a *sector* in the circos plot. Tracks can then be added using the coordinate within each sector (chromosome) provided and relecvant data for each value of the coordinates can be plotte for the respective track. Examples could be genes, transcripts, bases (SNPs, CpG sites) etc, each with start and the end site. 

The *ciclize* package in R and *circos* algorithm in perl provide easy infrastructure to plot geneomic data. 



Circular layout is very useful to represent complicated information. 
 - it elegantly represents information with long axes or a large amount of categories
 - it intuitively shows data with multiple tracks focusing on the same object
 - it easily demonstrates relations between elements
 - it os aesthetically pleasing way of representing data

Circos plots provide and excellent framework to visulaize multi assay omics data. The circular layout allows us to plot several layers of data based on coordinates. Genomic coordinates are one good example where the whole genome is displayed in the form of coordintes. The genomic coordinates can then be used to stack different layers of data.

A circular layout is composed of sectors and tracks. For data in different categories, they are allocated into different sectors, and for multiple measurements on the same category, they are represented as stacked tracks from outside of the circle to the inside. The intersection of a sector and a track is called a cell (or a grid, a panel), which is the basic unit in a circular layout. It is an imaginary plotting region for drawing data points.

# Principles of creating circos plot

The routine in making a circos plot works like

**initialize layout -> create track -> add graphics -> create track -> add graphics - ... -> clear**


## A basic example of circos plot

Let us create a small dataset and 

```{r}
set.seed(999)
n = 1000
df = data.frame(factors = sample(letters[1:8], n, replace = TRUE),
    x = rnorm(n), y = runif(n), z=rnorm(n))

head(df)
```


## Initialize the circular layout

```{r}
circos.par("track.height" = 0.1)
circos.initialize(factors = df$factors, x = df$x)
```

## create track

```{r}
circos.track(factors = df$factors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + mm_y(5), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})

circos.clear()
```

## add graphics


```{r}
circos.par("track.height" = 0.1)
circos.initialize(factors = df$factors, x = df$x)
circos.track(factors = df$factors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + mm_y(5), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})
col = rep(c("#FF0000", "#00FF00"), 4)
circos.trackPoints(df$factors, df$x, df$y, col = col, pch = 16, cex = 0.5)
circos.text(-1, 0.5, "text", sector.index = "a", track.index = 1)
```

## Add more tracks 

```{r}
bgcol = rep(c("#EFEFEF", "#CCCCCC"), 4)
circos.trackHist(df$factors, df$z, bin.size = 0.2, bg.col = bgcol, col = NA)
```




```{r}
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    breaks = seq(xlim[1], xlim[2], by = 0.1)
    n_breaks = length(breaks)
    circos.rect(breaks[-n_breaks], rep(ylim[1], n_breaks - 1),
                breaks[-1], rep(ylim[2], n_breaks - 1),
                col = rand_color(n_breaks), border = NA)
})

```

## Create links 

```{r}
circos.link("a", 0, "b", 0, h = 0.4)
circos.link("c", c(-0.5, 0.5), "d", c(-0.5,0.5), col = "red",
    border = "blue", h = 0.2)
circos.link("e", 0, "g", c(-1,1), col = "green", border = "black", lwd = 2, lty = 2)
```

```{r}
circos.clear()
```


A circular layout is composed of sectors and tracks. For data in different categories, they are allocated into different sectors, and for multiple measurements on the same category, they are represented as stacked tracks from outside of the circle to the inside. The intersection of a sector and a track is called a cell (or a grid, a panel), which is the basic unit in a circular layout. It is an imaginary plotting region for drawing data points.

*low-level graphic functions that can be used for adding graphics.*

 - circos.points() : adds points in a cell.
 - circos.lines() : adds lines in a cell.
 - circos.segments() : adds segments in a cell.
 - circos.rect()`: adds rectangles in a cell.
 - circos.polygon(): adds polygons in a cell.
 - circos.text(): adds text in a cell.
 - circos.axis() ands circos.yaxis(): add axis in a cell.

*Following function draws links between two positions in the circle:*

circos.link()


*Following functions draw high-level graphics:*

circos.barplot(): draw barplots.
circos.boxplot(): draw boxplots.
circos.violin(): draws violin plots.
circos.heatmap(): draw circular heatmaps.
circos.raster(): draw raster images.
circos.arrow(): draw circular arrows.

*Following functions arrange the circular layout.*

circos.initialize(): allocates sectors on the circle.
circos.track(): creates plotting regions for cells in one single track.
circos.update(): updates an existed cell.
circos.par(): graphic parameters.
circos.info(): prints general parameters of current circular plot.
circos.clear(): resets graphic parameters and internal variables.


## Overview of the functions in "circlize" package


## Applications in Genomics/omics

Let us creat a bed file using the generateRandomBed() function from the "circlize" package.

```{r}
set.seed(123)
bed = generateRandomBed(nc = 2, nr=500, fun = function(k) sample(letters, k, replace = TRUE))
head(bed)
```

All information in circlise is expected in the form of a dataframe.

```{r}
str(bed)
```


```{r}
circos.initializeWithIdeogram()
text(0, 0, "default", cex = 1)
```

```{r}
bed = generateRandomBed(nc = 2)
head(bed, n = 2)
circos.initializeWithIdeogram(plotType = NULL)
circos.genomicTrackPlotRegion(bed, panel.fun = function(region, value, ...) {
    if(CELL_META$sector.index == "chr1") {
        print(head(region, n = 2))
        print(head(value, n = 2))
    }
})
```

```{r}
circos.initializeWithIdeogram(species = "hg19")
circos.initializeWithIdeogram(species = "mm10")
```

```{r}
set.seed(123)
circos.initializeWithIdeogram(plotType = NULL)
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.rect(xlim[1], 0, xlim[2], 1, col = rand_color(1))
    circos.text(mean(xlim), mean(ylim), chr, cex = 0.7, col = "white",
        facing = "inside", niceFacing = TRUE)
}, track.height = 0.15, bg.border = NA)
```


```{r}
extend_chromosomes = function(bed, chromosome, prefix = "zoom_") {
    zoom_bed = bed[bed[[1]] %in% chromosome, , drop = FALSE]
    zoom_bed[[1]] = paste0(prefix, zoom_bed[[1]])
    rbind(bed, zoom_bed)
}
```


```{r}
cytoband = read.cytoband()
cytoband_df = cytoband$df
chromosome = cytoband$chromosome

xrange = c(cytoband$chr.len, cytoband$chr.len[c("chr1", "chr2")])
normal_chr_index = 1:24
zoomed_chr_index = 25:26

# normalize in normal chromsomes and zoomed chromosomes separately
sector.width = c(xrange[normal_chr_index] / sum(xrange[normal_chr_index]), 
                 xrange[zoomed_chr_index] / sum(xrange[zoomed_chr_index])) 
```


```{r}
circos.par(start.degree = 90)
circos.initializeWithIdeogram(extend_chromosomes(cytoband_df, c("chr1", "chr2")), 
    sector.width = sector.width)
```


```{r}
bed = generateRandomBed(500)
circos.genomicTrack(extend_chromosomes(bed, c("chr1", "chr2")),
    panel.fun = function(region, value, ...) {
        circos.genomicPoints(region, value, pch = 16, cex = 0.3)
})
```


```{r}
circos.link("chr1", get.cell.meta.data("cell.xlim", sector.index = "chr1"),
    "zoom_chr1", get.cell.meta.data("cell.xlim", sector.index = "zoom_chr1"),
    col = "#00000020", border = NA)
circos.clear()
```

```{r}
circos.initializeWithIdeogram(plotType = NULL)
circos.genomicTrackPlotRegion(bed, panel.fun = function(region, value, ...) {
    if(CELL_META$sector.index == "chr1") {
        print(head(region, n = 2))
        print(head(value, n = 2))
    }
})
```

```{r}
circos.genomicTrackPlotRegion(data, ylim = c(0, 1),
    panel.fun = function(region, value, ...) {
        circos.genomicPoints(region, value, ...)
})
circos.genomicTrackPlotRegion(data, numeric.column = c("value1", "value2"), 
    panel.fun = function(region, value, ...) {
        circos.genomicPoints(region, value, ...)
})
```

```{r}
set.seed(123)
bed1 = generateRandomBed(nr = 100)
bed1 = bed1[sample(nrow(bed1), 20), ]
bed2 = generateRandomBed(nr = 100)
bed2 = bed2[sample(nrow(bed2), 20), ]

circos.initializeWithIdeogram()
circos.genomicLink(bed1, bed2, col = rand_color(nrow(bed1), transparency = 0.5), 
 border = NA)
```


# Creat a circos plot for human metabolic network
## Create files/data from cricos plot

1) Create Cooridinates from HMR where pathways represent segments and number of reactions in each pathway indicates the length of the segment

2) Annotate genes and metabolites to the reaction coordinates on the pathways

3) Plot relevant data for genes/reactions/metbolites or pathways 



## 1) Convert human Human metabolic model to a network

```{r HMR_import}
getwd()
setwd("/Users/cob-aaf/Documents/GitHub/workshop_omics_integration/visualization")
path <-  "./Data/HMRdatabase2_00.xlsx"
HMR_DB <- path %>%  readxl::excel_sheets() %>% purrr::set_names() %>% purrr::map(readxl::read_xlsx, path = path ) 
HMR_DB <- HMR_DB %>% purrr::map(janitor::clean_names)


## Select the relevant data from the HMR

HMR_Network <- HMR_DB$RXNS %>% dplyr::select(rxnid, gene_association, subsystem, equation)
head(HMR_Network)
```

As you can see, there are multiple reactions, genes and metabolites associated with each subsystem(pathway).
Reactions are easy because they are unique and serve as an index for the subsystem's length. If were are to accomodate all genes and metabolites such that they occupy unique coordinates in our circos plots, we need to expand this data.

At the moment we have `r length(HMR_Network$RXNS)` reactions in the network. 

```{r expand1}
## Split the gene associated with each reactions into long vector

HMR_Network <- HMR_Network %>% mutate(gene_association = strsplit(as.character(gene_association), ";")) %>% unnest(gene_association)

head(HMR_Network)
```

Now that the genes accomodate a row in the data and diminesions have increased to `r dim(HMR_Network)[1]` rows.

```{r expand2}
## Split the reaction equation to get the metabolites associated with each reaction

HMR_Network <- HMR_Network %>% mutate(metabolite = strsplit(as.character(equation), split =  "<=> | [+] | =>"))  %>% unnest(cols = c(metabolite) ) %>% mutate(metabolite = trimws(metabolite))

head(HMR_Network)
```



```{r}
#HMR_long <- HMR_Network %>% select(gene_association, rxnid, metabolite, subsystem) %>%  gather( key = axix, value = node ,factor_key = TRUE )

#HMR_long2 <- HMR_Network %>%  gather( key = axix, value = node , c(gene_association, rxnid, metabolite, subsystem),factor_key = TRUE )
```

## Create coordinates for metabolic model pathways


One can use the circular layout to plot various kinds of data including networks as long as the the coordinate system for the data can be established. Here we will use *subsystems* in human metabolic network to set up the coordinates of the plot. 
We have created a *network* where all genes, reactions, metabolites and pathways are present. `r table(HMR_Network$subsystem)`.


Let us start by filtering some of the metabolic pathways listed on the network that may not be most relevant.

```{r HMR_filter}
rem <-c(HMR_DB$RXNS$subsystem[grep("Transport", HMR_DB$RXNS$subsystem , ignore.case = TRUE )],"Isolated", "isolated" , "Artificial reactions" , "Exchange reactions" , "Miscellaneous" , "Pool reactions"  )  

pws <- table(HMR_DB$RXNS$subsystem) < 70 
pws_remove <-c( pws[pws==TRUE] %>% names(), rem)

HMR_Network_small <- HMR_Network[which(!HMR_Network$subsystem %in% pws_remove) ,]
```


```{r HMR_filter_dims}

HMR_dims <- lapply(apply(HMR_Network_small, 2, unique), length)
HMR_dims

```


```{r HMR_long}
HMR_long <- HMR_Network_small %>% select(gene_association, rxnid, metabolite, subsystem) %>%  gather( key = axis, value = node ,factor_key = TRUE )
#test <- HMR_long %>% group_by(axis) %>% {rle(.$node)}
HMR_long2 <- HMR_Network_small    %>%  gather( key = axix, value = node , c(gene_association, rxnid, metabolite, subsystem),factor_key = TRUE )
```




```{r, pathway_coord}
pathway_coord <- HMR_Network_small %>% select(subsystem) %>% table() %>% as.data.frame() 
colnames(pathway_coord) <- c("name", "end")
#pathway_coord <- as.data.frame(table(HMR_Network$subsystem))
shortname <- paste0(c(1:length(pathway_coord$name)), rep("-",length(pathway_coord$name)) ,
    str_trunc(as.character(pathway_coord$name), width = 6, side = "right", ellipsis = "-"),
    str_trunc(as.character(pathway_coord$name), width = 5, side = "center", ellipsis = "-"),
    str_trunc(as.character(pathway_coord$name), width = 6, side = "left", ellipsis = "")
)
pathway_coord <- data.frame( name =pathway_coord$name, start= rep(1, HMR_dims$subsystem), end= pathway_coord$end)

head(pathway_coord)
```



```{r rxn_coord}
## Reaction coordinates
rxn_coord <- HMR_Network_small %>% select( rxnid, subsystem, gene_association, metabolite) %>% group_by(subsystem) %>% add_count(rxnid, name = "width_rxn_pw" ) %>% ungroup()
rxn_coord <- rxn_coord %>%  group_by(subsystem) %>% add_count(subsystem, name = "width_pw" ) %>% ungroup()


reaction_coord <-rxn_coord[!duplicated(rxn_coord$rxnid),]

#reaction_coord <- reaction_coord %>% group_by(subsystem, width_rxn_pw) %>% summarise_all(count)

reaction_coord <-reaction_coord %>% group_by(subsystem) %>% summarise(
    rxnid =rxnid,
    rxnend =cumsum(width_rxn_pw),
    rxnwidth = width_rxn_pw,
    gene = gene_association,
    metabolite = metabolite
  ) %>% ungroup()

reaction_ranges <- plyranges::as_iranges(reaction_coord, end = reaction_coord$rxnend, width =reaction_coord$rxnwidth) %>% as_granges( seqnames=reaction_coord$subsystem)
```




```{r}
circos.genomicInitialize(reaction_ranges, axis.labels.cex = 0.01, sector.names  = paste(rep( "pw", 22) ,c(1:22)))
 circos.track(ylim = c(0, 1), 
  bg.col = rand_color(22), 
  bg.border = NA, track.height = 0.05)
```


Tracks are created and graphics are added by circos.genomicTrackPlotRegions(), or the short version circos.genomicTrack(). In following examples, chromosome will be used as the genomic category, and we assume data is simply a data frame in BED format (where the first column is the chromosome name, the second and third column are start and end positions, and the following columns are associated values).

## Data to be plotted
Let us create some data values that we will be plotting tn the track

```{r}
reaction_ranges$pval <- sample(abs(rnorm(1000)), length(reaction_ranges), replace = T)
reaction_ranges$fc <- sample((rnorm(1000, 0)), length(reaction_ranges), replace = T)
```

## Points
```{r}
react_select <- as.data.frame(reaction_ranges[sample(500)])
react_select_bed <- react_select[,c()]
```


```{r}
circos.genomicPoints(react_select, react_select, numeric.column = 9)


circos.par("track.height" = 0.1)
circos.initialize(factors = pathway_coord$name, x = pathway_coord$end)
```



```{r}
circos.track(factors = df_reaction_ranges$factors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + mm_y(5), 
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.6)
})

```

```{r}
png(filename = "./Data/cric1.png", width = 5000, height = 5000)
circos.genomicInitialize(reaction_ranges ,  labels.cex = 1, sector.names =paste(rep( "pw", 22) ,c(1:22)), tickLabelsStartFromZero = F)
dev.off()
```

```{r}
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.rect(xlim[1], 0, xlim[2], 1, col = rand_color(1))
    circos.text(mean(xlim), mean(ylim), chr, cex = 0.7, col = "white",
                labels = paste(rep( "pw", 22) ,c(1:22)),
                facing = "inside", niceFacing = TRUE)
}, track.height = 0.1, bg.border = NA)

```


```{r}
circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    chr = CELL_META$sector.index
    xlim = CELL_META$xlim
    ylim = CELL_META$ylim
    circos.barplot(xlim[1], 0, xlim[2], 1, col = rand_color(1))
    circos.text(mean(xlim), mean(ylim), chr, cex = 0.7, col = "white",
                labels = paste(rep( "pw", 22) ,c(1:22)),
                facing = "inside", niceFacing = TRUE)
}, track.height = 0.1, bg.border = NA)

```



```{r}
png(filename = "./Data/cric1.png", width = 3000, height = 3000, res = 300)
circos.genomicInitialize(reaction_ranges, plotType = "labels", sector.names = rep("", length(shortname)), track.height = 0.2, labels.cex = 0.1)
#circos.initializeWithIdeogram(plotType = NULL)
circos.genomicLabels(reaction_ranges, labels.column = 1, side = "outside", cex = 1,niceFacing = TRUE,  
    col = as.numeric(seqnames(reaction_ranges)), line_col = as.numeric(seqnames(reaction_ranges)))
dev.off()
```


```{r}
circos.genomicInitialize(pathway_coord, track.height = 0.1)


circos.genomicLabels(pathway_coord, labels.column = 1, side = "outside",
    col = as.numeric(factor(pathway_coord[[1]])), line_col = as.numeric(factor(pathway_coord[[1]])))

circos.trackPlotRegion(ylim = c(0, 1), panel.fun = function(x, y) {
chr = get.cell.meta.data("sector.index")
xlim = get.cell.meta.data("xlim")
ylim = get.cell.meta.data("ylim")
circos.rect(xlim[1], 0.1, xlim[2], 0.5, col = rand_color(1)) 
circos.text(xlim, 0.9, chr, cex = 0.5, adj = 0.1,facing = "clockwise",
niceFacing = TRUE) }, bg.border = NA)
circos.clear()
```

```{r gene_coord}
## Gene coordinates
HMR_gene <- HMR_Network_small %>% select( gene_association, subsystem) %>% group_by(subsystem) %>% add_count(gene_association, name = "width_gene_pw" ) %>% ungroup()
HMR_gene <- HMR_gene %>%  group_by(subsystem) %>% add_count(subsystem, name = "width_pw" ) %>% ungroup()
gene_coord <- HMR_gene[!duplicated(HMR_gene$gene_association),]

#reaction_coord <- reaction_coord %>% group_by(subsystem, width_rxn_pw) %>% summarise_all(count)

gene_coord <-gene_coord %>% group_by(subsystem) %>% summarise(
    genid =gene_association,
    geneend =cumsum(width_gene_pw),
    genewidth = width_gene_pw,
  ) %>% ungroup()

gene_ranges <- plyranges::as_iranges(gene_coord, end = gene_coord$geneend, width =gene_coord$genewidth) %>% as_granges( seqnames=gene_coord$subsystem)

```


```{r met_coord}
## Metabolite coordinates
HMR_met <- HMR_Network_small %>% select( metabolite, subsystem) %>% group_by(subsystem) %>% add_count(metabolite, name = "width_met_pw" ) %>% ungroup()
HMR_met <- HMR_met %>%  group_by(subsystem) %>% add_count(subsystem, name = "width_pw" ) %>% ungroup()
met_coord <- HMR_met[!duplicated(HMR_met$metabolite),]

#reaction_coord <- reaction_coord %>% group_by(subsystem, width_rxn_pw) %>% summarise_all(count)

met_coord <-met_coord %>% group_by(subsystem) %>% summarise(
    genid =metabolite,
    metend =cumsum(width_met_pw),
    metwidth = width_met_pw,
  ) %>% ungroup()

met_ranges <- plyranges::as_iranges(met_coord, end = met_coord$metend, width =met_coord$metwidth) %>% as_granges( seqnames=met_coord$subsystem)

```


```{r}
path_coord <- union_ranges(gene_ranges, reaction_ranges) %>% as.data.frame()
```

```{r}
png(filename = "./Data/cric1.png", width = 5000, height = 5000)
circos.genomicInitialize(reaction_ranges ,  labels.cex = 0.8)
dev.off()
```




```{r}
HMR_long <- HMR_Network_small %>% select(gene_association, rxnid, metabolite, subsystem) %>%  gather( key = axix, value = node ,factor_key = TRUE )
HMR_long2 <- HMR_Network_small    %>%  gather( key = axix, value = node , c(gene_association, rxnid, metabolite, subsystem),factor_key = TRUE )
```

```{r}
### example code
#r <- HMR_Network_small %>% group_by(subsystem) %>% select(rxnid) %>% rle()

HMR_Network_small <- HMR_Network_small %>% group_by(subsystem) %>%  add_count(rxnid, name = "width_rxn_pw" ) %>% ungroup() # get number of reactions for each pathway
HMR_Network_small <- HMR_Network_small %>% group_by(rxnid) %>%  add_count(gene_association, name = "width_gene_rxn" ) %>% ungroup() # get number genes for each reaction
HMR_Network_small <- HMR_Network_small %>% group_by(rxnid) %>%  add_count(metabolite, name = "width_met_rxn" ) %>% ungroup() # get number of metabolites for each reaction

 HMR_Network_small$test <- HMR_Network_small %>% group_by(subsystem, rxnid) %>% group_size() %>% rep(., .) 

HMR_Network_small %>% rle(rxnid)


HMR_Network_small <- HMR_Network_small %>% group_by(subsystem) %>%  mutate(rxn_no = rle(rxnid)$lengths %>% {rep(seq(length(.)), .)}) %>% ungroup()

HMR_Network_small <-  group_by(subsystem) %>%  mutate(rxn_no = rle(rxnid)$lengths %>% {rep(seq(length(.)), .)}) %>% ungroup()

HMR_Network_small <- HMR_Network_small %>% group_by(subsystem,rxnid) %>% group_size()


df <- data.frame(start=HMR_Network_small$rxn_start,
                 width=HMR_Network_small$width_rxn_pw)
df$start1 <- df$start+df$width
# produces IRanges
rng <- df %>% as_iranges()
rng

grng <- df %>% transform(seqnames =HMR_Network_small$subsystem) %>% as_granges()


HMR_Network_small <- HMR_Network_small%>% group_by(subsystem) %>% mutate(rxn = rle(rxnid)$lengths %>% {rep(seq(length(.)), .)})

HMR_Network_small <- HMR_Network_small %>% group_by(subsystem) %>% mutate(met = rle(metabolite)$lengths %>% {rep(seq(length(.)), .)})

HMR_Network_small <- HMR_Network_small %>% mutate(pw = rle(subsystem)$lengths %>% {rep(seq(length(.)), .)})

HMR_Network_small <- HMR_Network_small %>% mutate(pwend = rle(subsystem)$lengths)

HMR_Network_small <- HMR_Network_small %>% group_by(subsystem) %>%  mutate(cumpw= cumsum(gene_end2)) %>% ungroup()
HMR_Network_small <- 
HMR_Network_small %>%  group_by(subsystem) %>% base::unique(gene_end2)

#
y <- HMR_Network_small    %>%  gather( key = xx, value = pw , c(gene_end2, subsystem),factor_key = TRUE )

HMR_Network_small <- 

    
    
```



```{r}
par(mar = c(1, 1, 1, 1))
circos.genomicInitialize(plotType = NULL) 
circos.trackPlotRegion(ylim = c(0, 1), panel.fun = function(x, y) {
chr = get.cell.meta.data("sector.index")
xlim = get.cell.meta.data("xlim")
ylim = get.cell.meta.data("ylim")
circos.rect(xlim[1], 0, xlim[2], 0.5, col = rand_color(1)) circos.text(mean(xlim), 0.9, chr, cex = 0.5, facing = "clockwise",
niceFacing = TRUE) }, bg.border = NA)
```




```{r}
tmp <- HMR_Network_small %>% split(HMR_Network_small$subsystem)  
ir1 <- IRanges(start=cumsum(HMR_Network_small$gene_end2), width=HMR_Network_small$gene_end2)
#
#
#
x <- HMR_Network_small %>% split(subsystem)
HMR_Network_small$gene_end2 %>% cumsum()
  

seq_along(HMR_Network_small$gene_end)


```

```{r}

IRanges(start = HMR_Network_small$rxnid, end = HMR_Network_small$gene_end2)
        
        
x <-HMR_Network_small %>% group_by(pw) 

rep(x, cumsum(x))


```

```{r}
x <- HMR_DB$METS[which(HMR_Network$metabolite %in% HMR_DB$METS$metid), ]

HMR_Network$metabolite %in% unlist(HMR_DB$METS$metname) %>% table()


keycol <- "axis"
valuecol <- "ID"
gathercols <- c("GENE ASSOCIATION", "SUBSYSTEM", "EQUATION")

HMR_long2 <- gather_(HMR_Network, keycol, valuecol, gathercols)


#Gene_pathway<- 
l<-HMR_Network$gene_association %>% str_split(pattern = "or|and")
names(l)<-HMR_Network$subsystem
l<-l[!is.na(l)]

l1<-rep(lapply(l, length))

df_2<-data.frame(trimws(unlist(l)), rep(names(l), l1)) 
names(df_2)<-c("g","pw")

no_reactions<-table(HMR_Network$subsystem)

#number of genes perreaction

no_of_genes<- df_2$pw %>% table()
for_genes<-df_2[order(df_2$pw),]
gen<-as.character(for_genes$g)
reactions_genes<-data.frame(no_reactions[names(no_of_genes)],no_of_genes)
colnames(reactions_genes)<-c("pw","no_rxns","pw1","ngenes")

reactions_genes$start<- rep(0, dim(reactions_genes)[1])
reactions_genes$chr<-rep("chr", 137)
reactions_genes$dash<-rep("-", 137)
reactions_genes$end<-(reactions_genes$ngenes*reactions_genes$no_rxns)



reactions_genes$pw1<-paste("pw", 1:137,sep = "")
reactions_genes$ID<-c(1:137)
reactions_genes$col<-rep("vvdred", 137)
reaction_gene_karyo<-reactions_genes %>% select(chr,dash,pw1,ID, start,end, col)
write_delim(unname(reaction_gene_karyo), file.path("~/circos/data/karyotype/", "HMR.karytype.txt"), col_names = FALSE, quote_escape = FALSE)
```


## 2) Reaction coordinates


```{r}
library(IRanges)
library(data.table)

HMR<-HMR_Network[HMR_Network$SUBSYSTEM %in% reactions_genes$pw,]

HMR<-HMR[order(HMR$SUBSYSTEM), c(1,3)]


x2<-split(HMR,HMR$SUBSYSTEM)
df_2$g<-as.character(df_2$g)
x<-split(df_2,df_2$pw)

function()


###################### Create coordinates for Reaction ########################

x1 <- HMR_Network_small%>%  ungroup() 


Reaction_IRanges<-list()
for (i in 1:137){
Reaction_IRanges[i] <- IRanges(start = c(1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]-reactions_genes$ngenes[i], 
                               end = (1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]
                          
)}


Reaction_IRanges<-list()
for (i in 1:137){
Reaction_IRanges[i] <- IRanges(start = c(1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]-reactions_genes$ngenes[i], 
                               end = (1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]
                          
)}

names(Reaction_IRanges)<-reaction_gene_karyo[,2] # Name for pathways


for (i in 1:length(Reaction_IRanges)){                ##Name each metabolite
     names(Reaction_IRanges[[i]])<-x2[[i]]$ID
}

List_Iranges_Reactions<-IRangesList(Reaction_IRanges)

reactions_data<-rbindlist(lapply(list(List_Iranges_Reactions),FUN = as.data.frame))

###################### Create coordinates for Genes ###########################

Genes_IRanges<-list()
for (i in 1:137){
Genes_IRanges[i]<-IRanges(start = c(1:reactions_genes$ngenes[i])*reactions_genes$no_rxns[i]-reactions_genes$no_rxns[i], end = (1:reactions_genes$ngenes[i])*reactions_genes$no_rxns[i])

 }
for (i in 1:length(Genes_IRanges)){
     names(Genes_IRanges[[i]])<-x[[i]]$g
}

names(Genes_IRanges)<-reaction_gene_karyo[,2]

List_Iranges_genes<-IRangesList(Genes_IRanges)
genes_data<-rbindlist(lapply(list(List_Iranges_genes),FUN = as.data.frame))


#########################

```


```{r}
circos.par("track.height" = 0.1)
circos.initialize(factors = df$factors, x = df$x)
circos.track(factors = df$factors, y = df$y,
    panel.fun = function(x, y) {
        circos.text(CELL_META$xcenter, CELL_META$cell.ylim[2] + mm_y(5),
            CELL_META$sector.index)
        circos.axis(labels.cex = 0.5)
    }
)
```


## Meta data files for circos

## We want to create output for circos plot

### Significant pathways (Karyotype plot or not)

### Coordinates of and metadata for reactions in significant pathways (-log10 P val up and down)

### Coordinates and metadata for genes in significant pathways (-log10 p vals, up and down, fold change)

### Significant overlaps for Links between pathways (gene based)

## 

```{r}

pathway_coord<-merge(pathways_sig, reactions_genes, by.x = "Name", by.y = "pw")
reaction_coord<-merge(Reactions_sig, reactions_data,by.x = "Name", by.y = "names" )
genes_sig<-as.data.frame(gene_stats)
genes_sig<-genes_sig[genes_sig$myPval<0.05,]

path_circ<-pathway_coord %>% select("pw1","start","end", "p (dist.dir.up)")
path_circ$`p (dist.dir.up)`<--log10(path_circ$`p (dist.dir.up)`)
path_circ$`p (dist.dir.up)`[path_circ$`p (dist.dir.up)`==Inf]<-16          

reac_sig <- reaction_coord %>% select("group_name","start","end", "p (dist.dir.up)")

reac_sig$`p (dist.dir.up)`<--log10(reac_sig$`p (dist.dir.up)`)
reac_sig$`p (dist.dir.up)`[reac_sig$`p (dist.dir.up)`==Inf]<-10 
reac_sig<-reac_sig[which(reac_sig$group_name %in% path_circ$pw1),]
#genes_coord<- cbind(genes_sig, genes_data[which(rownames(genes_sig) %in% genes_data$names), ])

```


```{r}
library(BioCircos)
genome_tmp<-reaction_gene_karyo[reaction_gene_karyo$pw1 %in% path_circ$pw1,]
genome <-as.list(genome_tmp$end)
names(genome)<-genome_tmp$pw1
track2<-BioCircosBarTrack(trackname = "pathway", chromosomes = path_circ$pw1, starts = path_circ$start, ends = path_circ$end-1,color = "blue", values = path_circ$`p (dist.dir.up)`)

track1<-BioCircosHeatmapTrack(trackname = "pathway", chromosomes = path_circ$pw1, starts = path_circ$start, ends = path_circ$end-1,color = c("red", "darkgreen"), values = path_circ$`p (dist.dir.up)`, minRadius = 0.60, maxRadius = 0.75)

track1<-track1 + BioCircosBarTrack(trackname = "reaction", chromosomes = reac_sig$group_name, starts = reac_sig$start, ends = reac_sig$end-1,color = "darkblue", values = reac_sig$`p (dist.dir.up)`, minRadius = 0.70, maxRadius = 0.90)


BioCircos(tracklist = track1,genome = genome, genomeFillColor = c("tomato4", "darkblue"),
  genomeTicksScale = 4e+3, genomeLabelOrientation = 90, height = '800px', width = '900px'
  )
```



### Hiveplot

Hive plots are an alternative way of visualizing the networks where different parameters and attributes of networks can be plotted on multiple linear axi. The attibutes of nodes and edges are very similar to a network except that instead of trying to intepret intesities of colors, degree of nodes etc. in a network, one can compare network paramaters on a linear scale. An advantage of hive plots for networks is that the subjective observations of nodes and edges of network layout can be obejectively compared.

It all sounds wordy but once you see your results in prtactice, it is much easier to understand.
The interpretation of values mapped to the axis can

1) Create Sif file
2) Create Gsc file 
### Perform reporter metabolites analyses

```{r}
library(HiveR)
library(tidyverse)
DE_genes_palm_controls_EdgeR <- read_csv("./Data/DE-genes-palm-controls-EdgeR.csv")

DE_genes_palm_controls_EdgeR<-DE_genes_palm_controls_EdgeR[!(is.na(DE_genes_palm_controls_EdgeR$`P-value`)),]
DE_genes_palm_controls_EdgeR<-as.data.frame(DE_genes_palm_controls_EdgeR)
rownames(DE_genes_palm_controls_EdgeR)<-DE_genes_palm_controls_EdgeR$`ENSG ID`

myPval<- DE_genes_palm_controls_EdgeR$`P-value`
names(myPval)<-DE_genes_palm_controls_EdgeR$`ENSG ID`
myFC<-DE_genes_palm_controls_EdgeR$`log2 FC`
names(myFC)<-DE_genes_palm_controls_EdgeR$`ENSG ID`

```

#Load the network data for pathways and metabolites

```{r}
Gene_stats<-read.delim("Data/GLS_Palm_vs_control_rep_pathway.txt")
met_stats<-read.delim("Data/GSS_Palm_vs_control_rep_met.txt")
path_stats<-  read.delim("Data/GSS_Palm_vs_control_rep_pathway.txt")
met_gene<-read.delim("Data/GSC_Palm_vs_control_rep_met.txt", header = FALSE)
path_gene<-read.delim("Data/GSC_Palm_vs_control_rep_pathway.txt", header = FALSE)
```


```{r}
hive1 <- edge2HPD(edge_df = met_gene)
```

```{r}
met_gene_z<- met_gene[which(met_gene$V1 %in% met_stats$Name),]
path_gene_z<-path_gene[which(as.character(path_gene$V1) %in% as.character(path_stats$Name)),]


Gene_stats$ID1<-c(1:(dim(Gene_stats)[1]))
Gene_stats$padj<-p.adjust(Gene_stats$p, method = "fdr")
met_gene$ID2<-as.integer(as.numeric(factor(met_gene$V1)))
path_gene$ID3<-as.integer(as.numeric(factor(path_gene$V1)))

gene_met_tmp<-merge(met_gene, Gene_stats, by.x = "V2", by.y ="g" )
gene_met_tmp$ID11<-as.integer(as.numeric(factor(gene_met_tmp$V2)))
gene_met_tmp$ID11<-gene_met_tmp$ID11+max(gene_met_tmp$ID2)
gene_met_tmp<-merge(gene_met_tmp, path_gene, by.x = "V2", by.y = "V2" )
gene_met_tmp$ID31<- gene_met_tmp$ID3+max(gene_met_tmp$ID11)
```

```{r}
met_stats_sig<-(met_stats[met_stats$p.adj..non.dir..<0.05,])
path_stats_sig<-path_stats[path_stats$p.adj..non.dir..<0.00001,]

sig_mets<-met_stats_sig$Name %>%as.character()
sig_paths<-path_stats_sig$Name%>% as.character()


gene_met_tmp_sig<-gene_met_tmp[which( as.character(gene_met_tmp$V1.x)%in% sig_mets|  as.character(gene_met_tmp$V1.y)%in% sig_paths),]

gene_met_tmp_sig <-gene_met_tmp_sig[gene_met_tmp_sig$padj<0.000001,]

#Create a node table
x_sig<-unique(c(gene_met_tmp_sig$ID2, gene_met_tmp_sig$ID11,gene_met_tmp_sig$ID31)) #nodes
x1_sig<-as.character((c(as.character(unique(gene_met_tmp_sig$V1.x)), as.character(unique(gene_met_tmp_sig$V2)),as.character(unique(gene_met_tmp_sig$V1.y))))) #labels
x2_sig<-rep(1:3, c(length(unique(gene_met_tmp_sig$V1.x)), length(unique(gene_met_tmp_sig$V2)),length(unique(gene_met_tmp_sig$V1.y)))) # assign axis

x3_sig<-  c(-log10(Gene_stats[which(x1_sig %in% Gene_stats$g),"padj"])*50, -log10(met_stats[which(x1_sig %in% met_stats$Name),"p..non.dir.."])*50,-log10(path_stats[which(path_stats$Name %in% x1_sig), "p..non.dir.."])*100 )   #radius i.e., postion on the axis is arranged by meanigful graphical parameter e.g., significance

x4_sig<-  scale(abs(c(Gene_stats[which(x1_sig %in% rownames(Gene_stats)),"myFC"], met_stats_sig[which(x1_sig %in% met_stats$Name ), "Genes..tot." ],path_stats[which(path_stats$Name %in% x1_sig), "Genes..tot." ] )), center = FALSE, scale = TRUE) #size of the node e.g., Degree of the node or fold change etc
x4_sig_a<-rep(1, length(x3_sig))

x5_sig<-  as.character(rep(c("#E41A1C" ,"#984EA3", "#4DAF4A"), c(length(unique(gene_met_tmp_sig$V1.x)), length(unique(gene_met_tmp_sig$V2)),length(unique(gene_met_tmp_sig$V1.y)))))   #Node color
  
#xx<-c(gene_stats[which(rownames(gene_stats) %in%x1),"myPval"], met_stats[which(met_stats$Name %in% x1), "p..non.dir.."],path_stats[which(path_stats$Name %in% x1), "p..non.dir.."] )
  

```

```{r}
# Create and edge table
y_sig<-  c(gene_met_tmp_sig$ID2, gene_met_tmp_sig$ID11,gene_met_tmp_sig$ID2 )#ID1
y1_sig<- c(gene_met_tmp_sig$ID11, gene_met_tmp_sig$ID31,gene_met_tmp_sig$ID31)#ID2
y2_sig<- rep(2, length(y_sig)) # weight
y3_sig<- as.character(rep ("pink" , length(y_sig)))#color



```


```{r}
HivePlotData<-list()
nodes_sig<-data.frame(x_sig,x1_sig,x2_sig,x3_sig,x4_sig_a,x5_sig)
colnames(nodes_sig)<-c("id", "lab", "axis","radius", "size", "color")
edges_sig<- data.frame(y_sig, y1_sig, y2_sig, y3_sig)
colnames(edges_sig)<- c("id1", "id2", "weight", "color")
type<- c("2D")
desc<-c("many nodes")
#axis.cols<- c("#E41A1C" ,"#377EB8", "#4DAF4A")
axis.cols<- c("blue" ,"purple", "green")

Data_hive_sig<- list(nodes_sig, edges_sig, type, desc, axis.cols)

names(Data_hive_sig)<-c("nodes", "edges", "type", "desc", "axis.cols")
Data_hive_sig$nodes$lab<-as.character(Data_hive_sig$nodes$lab)
Data_hive_sig$nodes$color<-as.character(Data_hive_sig$nodes$color)
Data_hive_sig$edges$color<-as.character(Data_hive_sig$edges$color)



class(Data_hive_sig)<-"HivePlotData"


```

```{r}
plotHive(Data_hive_sig, bkgnd = "black",
	axLabs = c("Metabolites", "Genes", "Pathways"),
	axLab.pos = c(1, 1),
	ch = 0.5, method = "norm"
	#axLab.gpar = gpar(fontsize = 5),
#	anNodes = "Data_hive_signodes.txt",
	#anNode.gpar = gpar(col = "black"),
#	grInfo = "Data_hive_siggraphics.txt"
	)
```


