{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "32a985e0-b190-45f5-833b-37e1f5b3b17d",
   "metadata": {},
   "source": [
    "# Topology lab, part 3\n",
    "\n",
    "## Functional analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2de2a2de-56ca-48f1-82e1-25d22b832f0f",
   "metadata": {},
   "source": [
    "In order to perform functional enrichment, we will extract each of the communities and perform a hypergeometric test **on the genes** to understand whether they are particularly enriched in specific biological functions.  \n",
    "We will use [enrichr](https://gseapy.readthedocs.io/en/master/gseapy_example.html#2.-Enrichr-Example) to perform the gene set enrichment analysis. As background we will use the full list of genes that were quantified.\n",
    "\n",
    "We will look at 3 gene set libraries. Should you have other kinds of data, enrichr allows you to define your own feature sets and perform a similar analysis. The challenge is in identifying comprehensive and well curated gene sets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "807100bd-159a-4681-9532-22286b61e5c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import gseapy as gp\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5d3f8892-7d57-4e99-b5e4-6e0d1d034325",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('data/serialization/degree_data.json', 'r') as file:\n",
    "    degree_data = json.load(file)\n",
    "\n",
    "feat_lists = pd.read_csv(\"data/serialization/feat_lists.csv\", sep = \"\\t\")\n",
    "comm_counts = pd.read_csv(\"data/serialization/comm_counts.csv\", sep = \"\\t\", index_col = 0)\n",
    "data=pd.read_csv('data/met_genes.tsv', sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1c45b7b9-49ff-4aa5-a2ed-f6d249294567",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HDSigDB_Human_2021', 'HumanCyc_2015', 'HumanCyc_2016', 'Human_Gene_Atlas', 'Human_Phenotype_Ontology', 'KEGG_2019_Human', 'KEGG_2021_Human', 'NIH_Funded_PIs_2017_Human_AutoRIF', 'NIH_Funded_PIs_2017_Human_GeneRIF', 'NURSA_Human_Endogenous_Complexome', 'RNAseq_Automatic_GEO_Signatures_Human_Down', 'RNAseq_Automatic_GEO_Signatures_Human_Up', 'Tissue_Protein_Expression_from_Human_Proteome_Map', 'WikiPathway_2021_Human', 'WikiPathway_2023_Human', 'WikiPathways_2019_Human', 'WikiPathways_2024_Human']\n",
      "['GO_Biological_Process_2013', 'GO_Biological_Process_2015', 'GO_Biological_Process_2017', 'GO_Biological_Process_2017b', 'GO_Biological_Process_2018', 'GO_Biological_Process_2021', 'GO_Biological_Process_2023']\n",
      "['KEGG_2013', 'KEGG_2015', 'KEGG_2016', 'KEGG_2019_Human', 'KEGG_2019_Mouse', 'KEGG_2021_Human']\n",
      "['OMIM_Disease']\n"
     ]
    }
   ],
   "source": [
    "# extract all the available human libraries in enrichr\n",
    "libraries = gp.get_library_name()\n",
    "print([lib for lib in libraries if 'Human' in lib])\n",
    "print([lib for lib in libraries if 'GO_Biological_Process' in lib])\n",
    "print([lib for lib in libraries if 'KEGG' in lib])\n",
    "print([lib for lib in libraries if 'OMIM_Disease' in lib])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b3087500-9fcd-4aba-8d05-6326e1b65822",
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analyzing pos network | Comm: c1/7  | BP: GO_Biological_Process_2023\n",
      "Error for: pos c1 GO_Biological_Process_2023\n",
      "Analyzing pos network | Comm: c1/7  | BP: KEGG_2021_Human\n",
      "Error for: pos c1 KEGG_2021_Human\n",
      "Analyzing pos network | Comm: c1/7  | BP: OMIM_Disease\n",
      "Error for: pos c1 OMIM_Disease\n",
      "Analyzing pos network | Comm: c2/7  | BP: GO_Biological_Process_2023\n",
      "Analyzing pos network | Comm: c2/7  | BP: KEGG_2021_Human\n",
      "Analyzing pos network | Comm: c2/7  | BP: OMIM_Disease\n",
      "Error for: pos c2 OMIM_Disease\n",
      "Analyzing pos network | Comm: c3/7  | BP: GO_Biological_Process_2023\n",
      "Analyzing pos network | Comm: c3/7  | BP: KEGG_2021_Human\n",
      "Analyzing pos network | Comm: c3/7  | BP: OMIM_Disease\n",
      "Error for: pos c3 OMIM_Disease\n",
      "Analyzing pos_w network | Comm: c1/7  | BP: GO_Biological_Process_2023\n",
      "Error for: pos_w c1 GO_Biological_Process_2023\n",
      "Analyzing pos_w network | Comm: c1/7  | BP: KEGG_2021_Human\n",
      "Error for: pos_w c1 KEGG_2021_Human\n",
      "Analyzing pos_w network | Comm: c1/7  | BP: OMIM_Disease\n",
      "Error for: pos_w c1 OMIM_Disease\n",
      "Analyzing pos_w network | Comm: c2/7  | BP: GO_Biological_Process_2023\n",
      "Analyzing pos_w network | Comm: c2/7  | BP: KEGG_2021_Human\n",
      "Analyzing pos_w network | Comm: c2/7  | BP: OMIM_Disease\n",
      "Error for: pos_w c2 OMIM_Disease\n",
      "Analyzing pos_w network | Comm: c3/7  | BP: GO_Biological_Process_2023\n",
      "Analyzing pos_w network | Comm: c3/7  | BP: KEGG_2021_Human\n",
      "Analyzing pos_w network | Comm: c3/7  | BP: OMIM_Disease\n",
      "Error for: pos_w c3 OMIM_Disease\n",
      "Analyzing all network | Comm: c1/6  | BP: GO_Biological_Process_2023\n",
      "Analyzing all network | Comm: c1/6  | BP: KEGG_2021_Human\n",
      "Analyzing all network | Comm: c1/6  | BP: OMIM_Disease\n",
      "Error for: all c1 OMIM_Disease\n",
      "Analyzing all network | Comm: c2/6  | BP: GO_Biological_Process_2023\n",
      "Analyzing all network | Comm: c2/6  | BP: KEGG_2021_Human\n",
      "Analyzing all network | Comm: c2/6  | BP: OMIM_Disease\n",
      "Error for: all c2 OMIM_Disease\n",
      "Analyzing all network | Comm: c3/6  | BP: GO_Biological_Process_2023\n",
      "Analyzing all network | Comm: c3/6  | BP: KEGG_2021_Human\n",
      "Analyzing all network | Comm: c3/6  | BP: OMIM_Disease\n",
      "Error for: all c3 OMIM_Disease\n",
      "Analyzing knn network | Comm: c1/5  | BP: GO_Biological_Process_2023\n",
      "Error for: knn c1 GO_Biological_Process_2023\n",
      "Analyzing knn network | Comm: c1/5  | BP: KEGG_2021_Human\n",
      "Error for: knn c1 KEGG_2021_Human\n",
      "Analyzing knn network | Comm: c1/5  | BP: OMIM_Disease\n",
      "Error for: knn c1 OMIM_Disease\n",
      "Analyzing knn network | Comm: c2/5  | BP: GO_Biological_Process_2023\n",
      "Analyzing knn network | Comm: c2/5  | BP: KEGG_2021_Human\n",
      "Analyzing knn network | Comm: c2/5  | BP: OMIM_Disease\n",
      "Error for: knn c2 OMIM_Disease\n",
      "Analyzing knn network | Comm: c3/5  | BP: GO_Biological_Process_2023\n",
      "Analyzing knn network | Comm: c3/5  | BP: KEGG_2021_Human\n",
      "Error for: knn c3 KEGG_2021_Human\n",
      "Analyzing knn network | Comm: c3/5  | BP: OMIM_Disease\n",
      "Error for: knn c3 OMIM_Disease\n",
      "Analyzing knn network | Comm: c4/5  | BP: GO_Biological_Process_2023\n",
      "Analyzing knn network | Comm: c4/5  | BP: KEGG_2021_Human\n",
      "Analyzing knn network | Comm: c4/5  | BP: OMIM_Disease\n",
      "Error for: knn c4 OMIM_Disease\n",
      "Analyzing knn network | Comm: c5/5  | BP: GO_Biological_Process_2023\n",
      "Analyzing knn network | Comm: c5/5  | BP: KEGG_2021_Human\n",
      "Analyzing knn network | Comm: c5/5  | BP: OMIM_Disease\n",
      "Error for: knn c5 OMIM_Disease\n"
     ]
    }
   ],
   "source": [
    "import ast\n",
    "import gseapy as gp\n",
    "\n",
    "# we will search 3 libraries for significantly enriched gene sets\n",
    "gene_sets = ['GO_Biological_Process_2023','KEGG_2021_Human','OMIM_Disease']\n",
    "background=[x for x in degree_data['all_u_names'] if x in data.loc[data.Type=='genes'].index]\n",
    "all_genes=data.loc[data.Type=='genes'].index\n",
    "\n",
    "def perform_enrich(network):\n",
    "    temp=feat_lists.copy()\n",
    "    temp=temp.loc[temp['network']==network]\n",
    "    output_enrichr=pd.DataFrame()\n",
    "    for comm in temp['community'].values:\n",
    "        # the dataframe holds the gene lists as strings, we parse them into lists\n",
    "        gl_string =temp.loc[temp['community']==comm, 'feat'].values[0]\n",
    "        gl = ast.literal_eval(gl_string)\n",
    "        gl=list([x for x in gl if x in all_genes])\n",
    "        # if the list is small we don't do the test\n",
    "        if(len(gl)<30):\n",
    "            continue\n",
    "        for bp in gene_sets:\n",
    "            print('Analyzing '+network+' network | Comm: '+comm+'/'+str(len(temp.index))+'  | BP: '+bp)\n",
    "            # When P-values are under thresholds the program exits\n",
    "            # We avoid this by using a try catch block\n",
    "            try:\n",
    "                enr=gp.enrichr(\n",
    "                    gene_list=gl,\n",
    "                    gene_sets=bp,\n",
    "                    background=background,\n",
    "                    outdir='Enrichr',\n",
    "                    format='png'\n",
    "                )\n",
    "                results=enr.results.sort_values('Adjusted P-value', ascending=True)\n",
    "                results=results.loc[results['Adjusted P-value']<0.05,]\n",
    "                results['BP']=bp\n",
    "                results['Comm']=comm\n",
    "                results['Graph']=network\n",
    "                output_enrichr=pd.concat([output_enrichr, results])\n",
    "            except Exception as e:\n",
    "                #print(f\"An unexpected error occurred: {e}\")\n",
    "                print(\"Error for:\", network, comm, bp)\n",
    "    return(output_enrichr)\n",
    "\n",
    "all_enriched=pd.DataFrame()\n",
    "#perform_enrich('pos')\n",
    "for net in ['pos', 'pos_w', 'all', 'knn']: \n",
    "    all_enriched=pd.concat([all_enriched,perform_enrich(net)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3d74f98-b97d-4f6f-8500-6416306cf814",
   "metadata": {},
   "source": [
    "From the output of the cell above you can readily see that some communities such as `c1` display no enriched terms from either [GO](http://geneontology.org/), [KEGG](https://www.genome.jp/kegg/) or [OMIM](https://www.omim.org/).\n",
    "\n",
    "Running the command above not only gives you the results after significance testing (Q<0.05), but it also outputs some preliminary barplots with the statistically significant results (found under `/Enrichr/`). For instance:\n",
    "\n",
    "> <img src=\"./Enrichr/GO_Biological_Process_2023.human.enrichr.reports.png\" style=\"width: 50%; height: auto;\" alt=\"GO BP 2023 stats\">\n",
    "> <img src=\"./Enrichr/KEGG_2021_Human.human.enrichr.reports.png\" style=\"width: 50%; height: auto;\" alt=\"KEGG 2021 stats\">"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88d49bd1-a188-4ccf-8c20-24680d8dd2e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_set</th>\n",
       "      <th>Term</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>BP</th>\n",
       "      <th>Comm</th>\n",
       "      <th>Graph</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Cytoplasmic Translation (GO:0002181)</td>\n",
       "      <td>6.596737e-21</td>\n",
       "      <td>1.768585e-17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10.248988</td>\n",
       "      <td>476.247014</td>\n",
       "      <td>EIF4A2;RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;...</td>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>c2</td>\n",
       "      <td>pos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Translation (GO:0006412)</td>\n",
       "      <td>2.464224e-20</td>\n",
       "      <td>3.303293e-17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.477737</td>\n",
       "      <td>382.768320</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...</td>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>c2</td>\n",
       "      <td>pos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Macromolecule Biosynthetic Process (GO:0009059)</td>\n",
       "      <td>7.129468e-20</td>\n",
       "      <td>6.371368e-17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.330309</td>\n",
       "      <td>367.262217</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...</td>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>c2</td>\n",
       "      <td>pos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Peptide Biosynthetic Process (GO:0043043)</td>\n",
       "      <td>2.043715e-19</td>\n",
       "      <td>1.369800e-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8.693117</td>\n",
       "      <td>374.102635</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...</td>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>c2</td>\n",
       "      <td>pos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>Gene Expression (GO:0010467)</td>\n",
       "      <td>3.477726e-19</td>\n",
       "      <td>1.864757e-16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6.725773</td>\n",
       "      <td>285.863776</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;H...</td>\n",
       "      <td>GO_Biological_Process_2023</td>\n",
       "      <td>c2</td>\n",
       "      <td>pos</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Gene_set  \\\n",
       "0  GO_Biological_Process_2023   \n",
       "1  GO_Biological_Process_2023   \n",
       "2  GO_Biological_Process_2023   \n",
       "3  GO_Biological_Process_2023   \n",
       "4  GO_Biological_Process_2023   \n",
       "\n",
       "                                              Term       P-value  \\\n",
       "0             Cytoplasmic Translation (GO:0002181)  6.596737e-21   \n",
       "1                         Translation (GO:0006412)  2.464224e-20   \n",
       "2  Macromolecule Biosynthetic Process (GO:0009059)  7.129468e-20   \n",
       "3        Peptide Biosynthetic Process (GO:0043043)  2.043715e-19   \n",
       "4                     Gene Expression (GO:0010467)  3.477726e-19   \n",
       "\n",
       "   Adjusted P-value  Old P-value  Old adjusted P-value  Odds Ratio  \\\n",
       "0      1.768585e-17            0                     0   10.248988   \n",
       "1      3.303293e-17            0                     0    8.477737   \n",
       "2      6.371368e-17            0                     0    8.330309   \n",
       "3      1.369800e-16            0                     0    8.693117   \n",
       "4      1.864757e-16            0                     0    6.725773   \n",
       "\n",
       "   Combined Score                                              Genes  \\\n",
       "0      476.247014  EIF4A2;RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;...   \n",
       "1      382.768320  RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...   \n",
       "2      367.262217  RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...   \n",
       "3      374.102635  RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;R...   \n",
       "4      285.863776  RPL4;RPL5;RPL30;RPL3;RPL32;RPL31;RPL34;RPLP1;H...   \n",
       "\n",
       "                           BP Comm Graph  \n",
       "0  GO_Biological_Process_2023   c2   pos  \n",
       "1  GO_Biological_Process_2023   c2   pos  \n",
       "2  GO_Biological_Process_2023   c2   pos  \n",
       "3  GO_Biological_Process_2023   c2   pos  \n",
       "4  GO_Biological_Process_2023   c2   pos  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_enriched.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8355ba33-6037-470f-b32c-ba9226bef98f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Graph</th>\n",
       "      <th>Comm</th>\n",
       "      <th>Term</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pos</td>\n",
       "      <td>c2</td>\n",
       "      <td>Cytoplasmic Translation (GO:0002181)</td>\n",
       "      <td>16.752374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pos</td>\n",
       "      <td>c2</td>\n",
       "      <td>Translation (GO:0006412)</td>\n",
       "      <td>16.481053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pos</td>\n",
       "      <td>c2</td>\n",
       "      <td>Macromolecule Biosynthetic Process (GO:0009059)</td>\n",
       "      <td>16.195767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pos</td>\n",
       "      <td>c2</td>\n",
       "      <td>Peptide Biosynthetic Process (GO:0043043)</td>\n",
       "      <td>15.863343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pos</td>\n",
       "      <td>c2</td>\n",
       "      <td>Gene Expression (GO:0010467)</td>\n",
       "      <td>15.729378</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Graph Comm                                             Term  \\\n",
       "0   pos   c2             Cytoplasmic Translation (GO:0002181)   \n",
       "1   pos   c2                         Translation (GO:0006412)   \n",
       "2   pos   c2  Macromolecule Biosynthetic Process (GO:0009059)   \n",
       "3   pos   c2        Peptide Biosynthetic Process (GO:0043043)   \n",
       "4   pos   c2                     Gene Expression (GO:0010467)   \n",
       "\n",
       "   Adjusted P-value  \n",
       "0         16.752374  \n",
       "1         16.481053  \n",
       "2         16.195767  \n",
       "3         15.863343  \n",
       "4         15.729378  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "enriched_terms=all_enriched.loc[:,['Graph','Comm','Term','Adjusted P-value']].copy()\n",
    "enriched_terms['Adjusted P-value']=-1*np.log10(enriched_terms['Adjusted P-value'])\n",
    "enriched_terms.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25665e6d-4d54-4329-b183-0ecedc33756c",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(7, 4))\n",
    "data_bars=pd.DataFrame(enriched_terms.groupby(['Graph','Comm'])['Term'].agg('count')).stack().reset_index().rename(columns={0:'Count'})\n",
    "sns.barplot(x='Graph', y='Count', data=data_bars, hue='Comm')\n",
    "ax.set_title('Number of significant Terms (Q < 0.05) per community')\n",
    "ax.legend(loc='right', bbox_to_anchor=(1.15, 1));\n",
    "plt.xticks(rotation=0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ac1e8e6-2b83-4058-9846-7351234d078f",
   "metadata": {},
   "source": [
    "Note that some of these communities are very big, which explains the big number of biological processes found above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a59b4831-feb7-4121-ac25-dc62a09f4139",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>c1</th>\n",
       "      <th>c2</th>\n",
       "      <th>c3</th>\n",
       "      <th>c4</th>\n",
       "      <th>c5</th>\n",
       "      <th>c6</th>\n",
       "      <th>c7</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>907.0</td>\n",
       "      <td>555.0</td>\n",
       "      <td>507.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos_w</th>\n",
       "      <td>902.0</td>\n",
       "      <td>556.0</td>\n",
       "      <td>506.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>all</th>\n",
       "      <td>696.0</td>\n",
       "      <td>657.0</td>\n",
       "      <td>611.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>knn</th>\n",
       "      <td>515.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>487.0</td>\n",
       "      <td>346.0</td>\n",
       "      <td>267.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          c1     c2     c3     c4     c5   c6   c7\n",
       "pos    907.0  555.0  507.0  119.0    9.0  3.0  2.0\n",
       "pos_w  902.0  556.0  506.0  122.0   11.0  3.0  2.0\n",
       "all    696.0  657.0  611.0  133.0    3.0  2.0  0.0\n",
       "knn    515.0  487.0  487.0  346.0  267.0  0.0  0.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###Number of genes/community\n",
    "# We skipped communities with <30 genes\n",
    "comm_counts.fillna(0).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7c7a28ff-4f92-46e7-90cb-4473a1887c83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Graph</th>\n",
       "      <th>Comm</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">all</th>\n",
       "      <th>c1</th>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c2</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c3</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">knn</th>\n",
       "      <th>c2</th>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c4</th>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c5</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">pos</th>\n",
       "      <th>c2</th>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c3</th>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">pos_w</th>\n",
       "      <th>c2</th>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c3</th>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Term\n",
       "Graph Comm      \n",
       "all   c1      66\n",
       "      c2      18\n",
       "      c3       3\n",
       "knn   c2      99\n",
       "      c3       1\n",
       "      c4     130\n",
       "      c5      15\n",
       "pos   c2      41\n",
       "      c3     132\n",
       "pos_w c2      41\n",
       "      c3     132"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(enriched_terms.groupby(['Graph','Comm'])['Term'].agg('count'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ac5efc7-beec-43d4-93bd-e8f8f2041a5d",
   "metadata": {},
   "source": [
    "We can now find whether the Full Network, Positively associated, and Positively associated weighted, show any common terms among their biggest communities. We do not compare with kNN-G as this shows very homogeneous and different communities than the other two networks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "74a6b6fa-a86a-4c41-abd6-9171e91be1cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Finding consensus\n",
    "temp=enriched_terms.copy()\n",
    "temp['comm_term']=temp.Comm+'_'+temp.Term\n",
    "temp=temp.loc[:,['Graph','comm_term']]\n",
    "\n",
    "consensus=pd.DataFrame()\n",
    "consensus=pd.concat([consensus, temp.loc[temp['Graph']=='pos']])\n",
    "consensus=pd.merge(consensus,\n",
    "                   temp.loc[temp['Graph']=='pos_w'], on=\"comm_term\", how='outer', suffixes=['pos','pos_w'])\n",
    "consensus=pd.merge(consensus, \n",
    "                   temp.loc[temp['Graph']=='all'], on=\"comm_term\", how='outer').rename(columns={'Graph':'all'})\n",
    "\n",
    "consensus=consensus.loc[consensus.isna().sum(1)==0].loc[:,['comm_term','Graphpos','Graphpos_w','all']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cedb1bcd-7b9d-4079-ba1b-1e22ddcba06f",
   "metadata": {},
   "source": [
    "Among the biggest communities we find several biological processes (55) that are simultaneously identified in the same community in the three graphs (`Full`, `Pos assoc`, and `Pos assoc weighted`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5329c661-90d6-4ec3-bccf-4478e0a69469",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "comm\n",
       "c2    18\n",
       "Name: comm_term, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus['comm']=[x[0] for x in consensus.comm_term.str.split('_')]\n",
    "consensus.groupby('comm')['comm_term'].agg('count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f4cd809a-b962-4bb2-bd12-53bfc54bd88c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>comm_term</th>\n",
       "      <th>Graphpos</th>\n",
       "      <th>Graphpos_w</th>\n",
       "      <th>all</th>\n",
       "      <th>comm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>c2_Circulatory System Development (GO:0072359)</td>\n",
       "      <td>pos</td>\n",
       "      <td>pos_w</td>\n",
       "      <td>all</td>\n",
       "      <td>c2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>c2_Coronavirus disease</td>\n",
       "      <td>pos</td>\n",
       "      <td>pos_w</td>\n",
       "      <td>all</td>\n",
       "      <td>c2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>c2_Cytoplasmic Translation (GO:0002181)</td>\n",
       "      <td>pos</td>\n",
       "      <td>pos_w</td>\n",
       "      <td>all</td>\n",
       "      <td>c2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>c2_ECM-receptor interaction</td>\n",
       "      <td>pos</td>\n",
       "      <td>pos_w</td>\n",
       "      <td>all</td>\n",
       "      <td>c2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>c2_Focal adhesion</td>\n",
       "      <td>pos</td>\n",
       "      <td>pos_w</td>\n",
       "      <td>all</td>\n",
       "      <td>c2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         comm_term Graphpos Graphpos_w  all  \\\n",
       "69  c2_Circulatory System Development (GO:0072359)      pos      pos_w  all   \n",
       "70                          c2_Coronavirus disease      pos      pos_w  all   \n",
       "71         c2_Cytoplasmic Translation (GO:0002181)      pos      pos_w  all   \n",
       "73                     c2_ECM-receptor interaction      pos      pos_w  all   \n",
       "74                               c2_Focal adhesion      pos      pos_w  all   \n",
       "\n",
       "   comm  \n",
       "69   c2  \n",
       "70   c2  \n",
       "71   c2  \n",
       "73   c2  \n",
       "74   c2  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "40f46462-7df7-46d6-ae80-29f315108574",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_terms.to_csv(\"data/serialization/enriched_terms.csv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95746b9-f812-4222-a3e7-8104936e04e6",
   "metadata": {},
   "source": [
    "**Questions**  \n",
    "- Would you exclude any communities based on its size?\n",
    "- Having identified these communities, how would you try to validate them?\n",
    "- Would you now determine the relevant community to investigate further?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8032ee30-adf6-4969-8f49-f75995c833cf",
   "metadata": {},
   "source": [
    "If you want to export communities to use them in other sections."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cc66bb2-527e-40bf-a708-f643157d0969",
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Compiles gene lists per community. We need Ensembl ids for further analyses\n",
    "# TODO: the code bellow is not working as it requires all three environments, or lots of files being saved\n",
    "# Homework?\n",
    "# Requires running the community detection from the previous section\n",
    "patlas=pd.read_csv('lab/data/proteinatlas.tsv', sep=\"\\t\").loc[:,['Ensembl','Gene']]\n",
    "def get_ensembl():\n",
    "    comm_counts=pd.DataFrame()\n",
    "    gene_lists=pd.DataFrame()\n",
    "    for i in [0,1,2,3]:\n",
    "        graph=[\"pos_w\",\"pos_w\",\"all_u\",\"knn\"][i]\n",
    "        comm=[pos_comm,pos_w_comm,all_comm,knn_comm][i]\n",
    "        name=['pos','pos_w','all','knn'][i]\n",
    "        temp=pd.DataFrame(list(zip(graph.vs['name'],[x+1 for x in comm.membership]))).rename(columns={0:'gene',1:'community'})\n",
    "\n",
    "        gl=pd.DataFrame(temp.groupby('community')['gene'].apply(list)).reset_index()\n",
    "        gl['community']=['c'+str(i) for i in gl['community']]\n",
    "        gl['network']=name\n",
    "        gl=gl.loc[:,['network','community','gene']]\n",
    "        gene_lists=pd.concat([gene_lists, gl])\n",
    "\n",
    "    gene_communities=gene_lists\n",
    "    gene_mat=pd.DataFrame()\n",
    "    for net in gene_communities['network'].unique():\n",
    "        temp=gene_communities.copy().loc[gene_communities['network']==net,]\n",
    "        for comm in temp['community'].unique():\n",
    "            gl=list(temp.copy().loc[temp['community']==comm,'gene'])[0]\n",
    "            el=[patlas.loc[patlas['Gene']==x,'Ensembl'].iloc[0] for x in gl if x in patlas['Gene'].values]\n",
    "\n",
    "            df=pd.DataFrame([net,comm,gl,el]).T\n",
    "            df.columns=['network','community','Gene','Ensembl']\n",
    "            gene_mat=pd.concat([gene_mat, df])\n",
    "                \n",
    "    return(gene_mat)\n",
    "\n",
    "get_ensembl().to_csv('data/gene_communities.tsv', sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2bb7494-6808-459c-b16c-04524f21fe77",
   "metadata": {},
   "source": [
    "# Conclusion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9692a3-2be5-46ee-bee3-c1d7fff22abc",
   "metadata": {},
   "source": [
    "Here we've performed some network analyses based on a met-gene association network. We've explored different centrality measures to characterize the networks, and identified the communities of genes in these networks. We have also used gene set enrichment analysis to characterize these communities based on the genes, but it remains to show whether similar results would be attained if we considered the metabolites in each community.  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
