---
title: "Tutorial"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
knitr::opts_chunk$set(echo = FALSE)
```


# Circos Plots

Circos plots provide and excellent framework to visulaize multi assay data. The circular layout allows us to plot several layers of data based on coordinates. Genomic coordinates are one good example where the whole genome is displayed in the form of coordintes. This can then be used to stack different layers of data.
This of course can be hacked to make plots for various different kinds of data. 
Here we will create coordinate for HMR and plot some of the data results we got from the Piano output. The coordinates from genomes are usually available from the gene annotation files but it requires a bit of thinking to create required data for non-standard visualizations.

## Create files/data from cricos plot

1) Karyotype (cooridinates) file from HMR taking into account the number of reactions and genes into account

2) Cooridinate files for Reactions and Genes

3) Meta data files
    a) Reaction statitics from Piano with coordinates
    b) Gene levels stats with coordinates
    


## 1) Karyotype/ coordinate file

```{r}
excelRxns <- read_delim("Data/excelRxns.txt",    "\t", escape_double = FALSE, trim_ws = TRUE)

HMR_Network<-excelRxns %>% dplyr::select(ID, `GENE ASSOCIATION`, SUBSYSTEM, EQUATION)

#Gene_pathway<- 
l<-HMR_Network$`GENE ASSOCIATION` %>% str_split(pattern = "or|and")
names(l)<-HMR_Network$SUBSYSTEM
l<-l[!is.na(l)]

l1<-rep(lapply(l, length))

df_2<-data.frame(trimws(unlist(l)), rep(names(l), l1)) 
names(df_2)<-c("g","pw")

no_reactions<-table(HMR_Network$SUBSYSTEM)

#number of genes perreaction

no_of_genes<- df_2$pw %>% table()
for_genes<-df_2[order(df_2$pw),]
gen<-as.character(for_genes$g)
reactions_genes<-data.frame(no_reactions[names(no_of_genes)],no_of_genes)
colnames(reactions_genes)<-c("pw","no_rxns","pw1","ngenes")

reactions_genes$start<- rep(0, dim(reactions_genes)[1])
reactions_genes$chr<-rep("chr", 137)
reactions_genes$dash<-rep("-", 137)
reactions_genes$end<-(reactions_genes$ngenes*reactions_genes$no_rxns)



reactions_genes$pw1<-paste("pw", 1:137,sep = "")
reactions_genes$ID<-c(1:137)
reactions_genes$col<-rep("vvdred", 137)
reaction_gene_karyo<-reactions_genes %>% select(chr,dash,pw1,ID, start,end, col)
write_delim(unname(reaction_gene_karyo), file.path("~/circos/data/karyotype/", "HMR.karytype.txt"), col_names = FALSE, quote_escape = FALSE)
```


## 2) Reaction coordinates

```{r}
library(IRanges)

library(data.table)
HMR<-HMR_Network[HMR_Network$SUBSYSTEM %in% reactions_genes$pw,]
HMR<-HMR[order(HMR$SUBSYSTEM), c(1,3)]
x2<-split(HMR,HMR$SUBSYSTEM)
df_2$g<-as.character(df_2$g)
x<-split(df_2,df_2$pw)


###################### Create coordinates for Reaction ########################

Reaction_IRanges<-list()
for (i in 1:137){
Reaction_IRanges[i]<-IRanges(start = c(1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]-reactions_genes$ngenes[i], end = (1:reactions_genes$no_rxns[i])*reactions_genes$ngenes[i]
                          
)}

names(Reaction_IRanges)<-reaction_gene_karyo[,2] # Name for pathways


for (i in 1:length(Reaction_IRanges)){                ##Name each metabolite
     names(Reaction_IRanges[[i]])<-x2[[i]]$ID
}

List_Iranges_Reactions<-IRangesList(Reaction_IRanges)

reactions_data<-rbindlist(lapply(list(List_Iranges_Reactions),FUN = as.data.frame))

###################### Create coordinates for Genes ###########################

Genes_IRanges<-list()
for (i in 1:137){
Genes_IRanges[i]<-IRanges(start = c(1:reactions_genes$ngenes[i])*reactions_genes$no_rxns[i]-reactions_genes$no_rxns[i], end = (1:reactions_genes$ngenes[i])*reactions_genes$no_rxns[i])

 }
for (i in 1:length(Genes_IRanges)){
     names(Genes_IRanges[[i]])<-x[[i]]$g
}

names(Genes_IRanges)<-reaction_gene_karyo[,2]

List_Iranges_genes<-IRangesList(Genes_IRanges)
genes_data<-rbindlist(lapply(list(List_Iranges_genes),FUN = as.data.frame))


#########################

```


## Meta data files for circos

## We want to create output for circos plot

### Significant pathways (Karyotype plot or not)

### Coordinates of and metadata for reactions in significant pathways (-log10 P val up and down)

### Coordinates and metadata for genes in significant pathways (-log10 p vals, up and down, fold change)

### Significant overlaps for Links between pathways (gene based)

## 

```{r}
pathway_coord<-merge(pathways_sig, reactions_genes, by.x = "Name", by.y = "pw")
reaction_coord<-merge(Reactions_sig, reactions_data,by.x = "Name", by.y = "names" )
genes_sig<-as.data.frame(gene_stats)
genes_sig<-genes_sig[genes_sig$myPval<0.05,]

path_circ<-pathway_coord %>% select("pw1","start","end", "p (dist.dir.up)")
path_circ$`p (dist.dir.up)`<--log10(path_circ$`p (dist.dir.up)`)
path_circ$`p (dist.dir.up)`[path_circ$`p (dist.dir.up)`==Inf]<-16          

reac_sig <- reaction_coord %>% select("group_name","start","end", "p (dist.dir.up)")

reac_sig$`p (dist.dir.up)`<--log10(reac_sig$`p (dist.dir.up)`)
reac_sig$`p (dist.dir.up)`[reac_sig$`p (dist.dir.up)`==Inf]<-10 
reac_sig<-reac_sig[which(reac_sig$group_name %in% path_circ$pw1),]
#genes_coord<- cbind(genes_sig, genes_data[which(rownames(genes_sig) %in% genes_data$names), ])

```


```{r}
library(BioCircos)
genome_tmp<-reaction_gene_karyo[reaction_gene_karyo$pw1 %in% path_circ$pw1,]
genome <-as.list(genome_tmp$end)
names(genome)<-genome_tmp$pw1
track2<-BioCircosBarTrack(trackname = "pathway", chromosomes = path_circ$pw1, starts = path_circ$start, ends = path_circ$end-1,color = "blue", values = path_circ$`p (dist.dir.up)`)

track1<-BioCircosHeatmapTrack(trackname = "pathway", chromosomes = path_circ$pw1, starts = path_circ$start, ends = path_circ$end-1,color = c("red", "darkgreen"), values = path_circ$`p (dist.dir.up)`, minRadius = 0.60, maxRadius = 0.75)

track1<-track1 + BioCircosBarTrack(trackname = "reaction", chromosomes = reac_sig$group_name, starts = reac_sig$start, ends = reac_sig$end-1,color = "darkblue", values = reac_sig$`p (dist.dir.up)`, minRadius = 0.70, maxRadius = 0.90)


BioCircos(tracklist = track1,genome = genome, genomeFillColor = c("tomato4", "darkblue"),
  genomeTicksScale = 4e+3, genomeLabelOrientation = 90, height = '800px', width = '900px'
  )
```



### Hiveplot

Hive plots are an alternative way of visualizing the networks where different parameters and attributes of networks can be plotted on multiple linear axi. The attibutes of nodes and edges are very similar to a network except that instead of trying to intepret intesities of colors, degree of nodes etc. in a network, one can compare network paramaters on a linear scale. An advantage of hive plots for networks is that the subjective observations of nodes and edges of network layout can be obejectively compared.

It all sounds wordy but once you see your results in prtactice, it is much easier to understand.
The interpretation of values mapped to the axis can

1) Create Sif file
2) Create Gsc file 
### Perform reporter metabolites analyses

```{r}
library(HiveR)
library(tidyverse)
DE_genes_palm_controls_EdgeR <- read_csv("Data/DE-genes-palm-controls-EdgeR.csv")

DE_genes_palm_controls_EdgeR<-DE_genes_palm_controls_EdgeR[!(is.na(DE_genes_palm_controls_EdgeR$`P-value`)),]
DE_genes_palm_controls_EdgeR<-as.data.frame(DE_genes_palm_controls_EdgeR)
rownames(DE_genes_palm_controls_EdgeR)<-DE_genes_palm_controls_EdgeR$`ENSG ID`

myPval<- DE_genes_palm_controls_EdgeR$`P-value`
names(myPval)<-DE_genes_palm_controls_EdgeR$`ENSG ID`
myFC<-DE_genes_palm_controls_EdgeR$`log2 FC`
names(myFC)<-DE_genes_palm_controls_EdgeR$`ENSG ID`

```

#Load the network data for pathwas and metabolites

```{r}
Gene_stats<-read.delim("Data/GLS_Palm_vs_control_rep_pathway.txt")
met_stats<-read.delim("Data/GSS_Palm_vs_control_rep_met.txt")
path_stats<-  read.delim("Data/GSS_Palm_vs_control_rep_pathway.txt")
met_gene<-read.delim("Data/GSC_Palm_vs_control_rep_met.txt", header = FALSE)
path_gene<-read.delim("Data/GSC_Palm_vs_control_rep_pathway.txt", header = FALSE)
```

```{r}
met_gene_z<- met_gene[which(met_gene$V1 %in% met_stats$Name),]
path_gene_z<-path_gene[which(as.character(path_gene$V1) %in% as.character(path_stats$Name)),]
Gene_stats$ID1<-c(1:(dim(Gene_stats)[1]))
Gene_stats$padj<-p.adjust(Gene_stats$p, method = "fdr")
met_gene$ID2<-as.integer(as.numeric(met_gene$V1))
path_gene$ID3<-as.integer(as.numeric(path_gene$V1))
gene_met_tmp<-merge(met_gene, Gene_stats, by.x = "V2", by.y ="g" )
gene_met_tmp$ID11<-as.integer(as.numeric(gene_met_tmp$V2))
gene_met_tmp$ID11<-gene_met_tmp$ID11+max(gene_met_tmp$ID2)
gene_met_tmp<-merge(gene_met_tmp, path_gene, by.x = "V2", by.y = "V2" )
gene_met_tmp$ID31<- gene_met_tmp$ID3+max(gene_met_tmp$ID11)
```

```{r}
met_stats_sig<-(met_stats[met_stats$p.adj..non.dir..<0.05,])
path_stats_sig<-path_stats[path_stats$p.adj..non.dir..<0.00001,]

sig_mets<-met_stats_sig$Name %>%as.character()
sig_paths<-path_stats_sig$Name%>% as.character()


gene_met_tmp_sig<-gene_met_tmp[which( as.character(gene_met_tmp$V1.x)%in% sig_mets|  as.character(gene_met_tmp$V1.y)%in% sig_paths),]

gene_met_tmp_sig <-gene_met_tmp_sig[gene_met_tmp_sig$padj<0.000001,]

#Create a node table
x_sig<-unique(c(gene_met_tmp_sig$ID2, gene_met_tmp_sig$ID11,gene_met_tmp_sig$ID31)) #nodes
x1_sig<-as.character((c(as.character(unique(gene_met_tmp_sig$V1.x)), as.character(unique(gene_met_tmp_sig$V2)),as.character(unique(gene_met_tmp_sig$V1.y))))) #labels
x2_sig<-rep(1:3, c(length(unique(gene_met_tmp_sig$V1.x)), length(unique(gene_met_tmp_sig$V2)),length(unique(gene_met_tmp_sig$V1.y)))) # assign axis

x3_sig<-  c(-log10(Gene_stats[which(x1_sig %in% Gene_stats$g),"padj"])*50, -log10(met_stats[which(x1_sig %in% met_stats$Name),"p..non.dir.."])*50,-log10(path_stats[which(path_stats$Name %in% x1_sig), "p..non.dir.."])*100 )   #radius i.e., postion on the axis is arranged by meanigful graphical parameter e.g., significance

x4_sig<-  scale(abs(c(Gene_stats[which(x1_sig %in% rownames(Gene_stats)),"myFC"], met_stats_sig[which(x1_sig %in% met_stats$Name ), "Genes..tot." ],path_stats[which(path_stats$Name %in% x1_sig), "Genes..tot." ] )), center = FALSE, scale = TRUE) #size of the node e.g., Degree of the node or fold change etc
x4_sig_a<-rep(1, length(x3_sig))

x5_sig<-  as.character(rep(c("#E41A1C" ,"#984EA3", "#4DAF4A"), c(length(unique(gene_met_tmp_sig$V1.x)), length(unique(gene_met_tmp_sig$V2)),length(unique(gene_met_tmp_sig$V1.y)))))   #Node color
  
#xx<-c(gene_stats[which(rownames(gene_stats) %in%x1),"myPval"], met_stats[which(met_stats$Name %in% x1), "p..non.dir.."],path_stats[which(path_stats$Name %in% x1), "p..non.dir.."] )
  

```

```{r}
# Create and edge table
y_sig<-  c(gene_met_tmp_sig$ID2, gene_met_tmp_sig$ID11,gene_met_tmp_sig$ID2 )#ID1
y1_sig<- c(gene_met_tmp_sig$ID11, gene_met_tmp_sig$ID31,gene_met_tmp_sig$ID31)#ID2
y2_sig<- rep(2, length(y_sig)) # weight
y3_sig<- as.character(rep ("pink" , length(y_sig)))#color



```


```{r}
HivePlotData<-list()
nodes_sig<-data.frame(x_sig,x1_sig,x2_sig,x3_sig,x4_sig_a,x5_sig)
colnames(nodes_sig)<-c("id", "lab", "axis","radius", "size", "color")
edges_sig<- data.frame(y_sig, y1_sig, y2_sig, y3_sig)
colnames(edges_sig)<- c("id1", "id2", "weight", "color")
type<- c("2D")
desc<-c("many nodes")
#axis.cols<- c("#E41A1C" ,"#377EB8", "#4DAF4A")
axis.cols<- c("blue" ,"purple", "green")

Data_hive_sig<- list(nodes_sig, edges_sig, type, desc, axis.cols)

names(Data_hive_sig)<-c("nodes", "edges", "type", "desc", "axis.cols")
Data_hive_sig$nodes$lab<-as.character(Data_hive_sig$nodes$lab)
Data_hive_sig$nodes$color<-as.character(Data_hive_sig$nodes$color)
Data_hive_sig$edges$color<-as.character(Data_hive_sig$edges$color)



class(Data_hive_sig)<-"HivePlotData"


```

```{r}
plotHive(Data_hive_sig, bkgnd = "black",
	axLabs = c("Metabolites", "Genes", "Pathways"),
	axLab.pos = c(1, 1),
	#axLab.gpar = gpar(fontsize = 5),
	anNodes = "Data_hive_signodes.txt",
	#anNode.gpar = gpar(col = "black"),
	grInfo = "Data_hive_siggraphics.txt"
	)
```


