<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="progressive" content="false" />
<meta name="allow-skip" content="false" />

<title>Tutorial</title>


<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>

<!-- taken from https://github.com/rstudio/rmarkdown/blob/67b7f5fc779e4cfdfd0f021d3d7745b6b6e17149/inst/rmd/h/default.html#L296-L362 -->
<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>
<!-- end tabsets -->



</head>

<body>



<div class="pageContent band">
<div class="bandContent page">

<div class="topics">

<div id="section-installation-and-setup" class="section level2">
<h2>Installation and setup</h2>
<p>Before we begin reading this tutorial and running the code, let us set up our R environment for this session. We will be using “renv” package in R to install all the necessary packages needed for the sessions.</p>
<blockquote>
<p><em>Note: Due to the use of the development versions of R packages that are not available through conda repositories we are not able to generate a conda environment for this session and an alternative solution needs to be used.</em></p>
</blockquote>
<pre class="r"><code>setwd(&quot;~/Documents/GitHub/workshop_omics_integration/session_meta/&quot;)
getwd()</code></pre>
<pre><code>## [1] &quot;/Users/cob-aaf/Documents/GitHub/workshop_omics_integration/session_meta&quot;</code></pre>
<pre class="r"><code>## make sure that the renv.lock file for the session exists and do following.
install.packages(&quot;renv&quot;)
renv::init()</code></pre>
<blockquote>
<p>If XML lib fails in ubuntu and you get the error “Error installing package ‘XML’:” go to terminal and install the libraries. “sudo apt-get update” followed by “sudo apt-get install libxml2-dev” Now, try the renv::init() again.</p>
</blockquote>
</div>
<div id="section-introduction" class="section level2">
<h2>Introduction</h2>
<p>Meta analyses are commonly used in clinical studies to assess the effect of a treatment or a genetic locus on a phenotype. With the advent of GWAS (Genome wide association studies), various studies report summary statistics on associated effects of genetic loci on phenotype. Meta analyses can be useful in determining whether the effect size is consistent across the body of data</p>
<p>The goal of a synthesis is to understand the results of any study in the context of all the other studies. First, we need to know whether or not the effect size is consistent across the body of data. If it is consistent, then we want to estimate the effect size as accurately as possible and to report that it is robust across the kinds of studies included in the synthesis. On the other hand, if it varies substantially from study to study, we want to quantify the extent of the variance and consider the implications. Meta-analysis is able to address these issues whereas the narrative review is not. We start with an example to show how meta-analysis and narrative review would approach the same question, and then use this example to highlight the key differences between the two.</p>
</div>
<div id="section-meta-analyses-methods" class="section level2">
<h2>Meta analyses Methods</h2>
<ol style="list-style-type: decimal">
<li><strong>p-value</strong></li>
</ol>
<!-- -->
<ol style="list-style-type: lower-alpha">
<li><a href="https://link.springer.com/chapter/10.1007%2F978-1-4612-4380-9_6"><strong>Fisher</strong></a>: Sum of minus log-transformed <em>P</em>-values where larger Fisher score reflects stronger aggregated differential expression evidence.</li>
<li><a href="https://cutt.ly/rc53t31"><strong>Stouffer</strong></a>: Sum of inverse normal transformed <em>P</em>-values where larger Stouffer score to reflect stronger aggregated statistical evidence.</li>
<li><a href="https://doi.org/10.1214/10-AOAS393"><strong>adaptively weighted Fisher(AW)</strong></a><strong>, <a href="https://ui.adsabs.harvard.edu/abs/2011arXiv1108.3180L/abstract">original publication</a></strong> : assigns different weights to each individual study  and it searches through all possible weights to find the best adaptive weight with the smallest derived <em>p</em>-value. One significant advantage of this method is its ability to indicate which studies contribute to the evidence aggregation and elucidates heterogeneity in the meta-analysis.</li>
<li><a href="https://psycnet.apa.org/record/1951-06623-001"><strong>minimum p-value (minP)</strong></a><strong>:</strong> The minP method takes the minimum <em>p</em>-value among the <em>K</em> studies as the test statistic</li>
<li><a href="https://psycnet.apa.org/record/1951-06623-001"><strong>maximum p-value (maxP)</strong></a><strong>:</strong> The maxP method takes maximum <em>p</em>-value as the test statistic</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/25383132/"><strong>rth ordered p-value (rOP)</strong></a><strong>:</strong> The rOP method takes the <em>r-</em>th order statistic among sorted <em>p</em>-values of <em>K</em>combined studies</li>
</ol>
<blockquote>
<p>Note: The assumption of uniformly distributed <em>P</em>-values under the null hypothesis or can be done non-parametrically by permutation-based analysis</p>
</blockquote>
<!-- -->
<ol start="2" style="list-style-type: decimal">
<li><strong>Effect Size based</strong></li>
</ol>
<!-- -->
<ol style="list-style-type: lower-alpha">
<li><a href="https://pubmed.ncbi.nlm.nih.gov/12855442/"><strong>fixed effects model (FEM)</strong></a>:FEM combines the effect size across <em>K</em> studies by assuming a simple linear model with an underlying true effect size plus a random error in each study</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/12855442/"><strong>random effects model (REM)</strong></a>: REM extends FEM by allowing random effects for the inter-study heterogeneity in the model. </li>
</ol>
<!-- -->
<ol start="3" style="list-style-type: decimal">
<li><strong>Rank based</strong></li>
</ol>
<!-- -->
<ol style="list-style-type: lower-alpha">
<li><a href="https://pubmed.ncbi.nlm.nih.gov/16982708/"><strong>rank product (rankProd)</strong></a>RankProd and RankSum are based on the common biological belief that if a gene is repeatedly at the top of the lists ordered by up- or down-regulation fold change in replicate experiments, the gene is more likely a DE gene.</li>
<li><a href="https://pubmed.ncbi.nlm.nih.gov/16982708/"><strong>naive sum of ranks and naive product of ranks</strong></a><strong>:</strong> These two methods apply a naïve product or sum of the DE evidence ranks across studies.</li>
</ol>
<div id="section-statistical-considerations" class="section level3">
<h3>Statistical considerations</h3>
<p>In addition to statistical methods mentioned above, a number of factors need to be considered for the choice of method. Different test statistics maybe used depending on the type of outcome variable (e.g. t-statistic or moderated t-statistic for binary outcome, F-statistic for multi-class outcome, regression or correlation coefficient for continuous outcome and log-rank statistic for survival outcome).</p>
<p>Here we will be using <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3463115/">MetaDE</a> package that has implemented above mentioned methods.</p>
<p>A summary of the methods and their implementations for respective outcome variables are listed in the table below as presented in the original publication.</p>
<p><img src="images/Statistical_methods_MetaDE.png" /></p>
<p>For a detailed review of the applied methods for meta analyses, related benchmarks and</p>
</div>
<div id="section-data-types" class="section level3">
<h3>Data types</h3>
</div>
<div id="section-required-packages" class="section level3">
<h3>Required packages</h3>
<pre class="r"><code>library(magrittr)
library(plyr)
library(preproc)
library(MetaQC)
library(MetaDE)
library(MetaPath)
library(MetaDCN)</code></pre>
</div>
</div>
<div id="section-summary-of-the-session" class="section level2">
<h2>Summary of the session</h2>
<p>I this session we will be working with the “prostate8.rda” data set that is located in the data directory of the “session_meta” directory of the github repository for the course.</p>
<p>Some of the R packages packages we are using here are not available through conda repositories and therefore need to be install from the github repositories for the respective packages</p>
<p>The session comprises of following main steps</p>
<ol style="list-style-type: decimal">
<li>Explore the data and the packages</li>
<li>Filter the data based on QC</li>
<li>Perform differential expression analyses and meta analyses</li>
<li>Perfrom pathway analyses for meta analyses results</li>
<li>Perfrom pathway meta analyses</li>
<li>Network meta analyses (optional)</li>
<li>Exercise: compare different meta analyses methods including AW-Fisher, REM, FEM ## Perform QC</li>
</ol>
</div>
<div id="section-data-for-the-session" class="section level2">
<h2>Data for the session</h2>
<p>We will using prostate dataset</p>
<p>Load the protaste data set</p>
<pre class="r"><code>load(file = &quot;./data/prostate8.rda&quot;)</code></pre>
<p>The prostate data is comprised of 8 microarray studies from different microarray platforms. Here it is provided in a list format where data matrices and corresponding labels are provided.</p>
<p>Let us take a quick look at the data.</p>
<pre class="r"><code>names(prostate8)</code></pre>
<pre><code>## [1] &quot;data&quot;      &quot;dataLabel&quot;</code></pre>
<p>We can have a look at the structure of each data sets and evaluate what sort of filtering steps we need to take.</p>
<pre class="r"><code>library(magrittr)
kableExtra::kable(lapply(prostate8$data, dim) %&gt;% as.data.frame(row.names = c(&quot;genes&quot;, &quot;samples&quot;)), booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Welsh
</th>
<th style="text-align:right;">
Yu
</th>
<th style="text-align:right;">
Lapointe
</th>
<th style="text-align:right;">
Varambally
</th>
<th style="text-align:right;">
Singh
</th>
<th style="text-align:right;">
Wallace
</th>
<th style="text-align:right;">
Nanni
</th>
<th style="text-align:right;">
Tomlins
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
genes
</td>
<td style="text-align:right;">
8798
</td>
<td style="text-align:right;">
8799
</td>
<td style="text-align:right;">
13579
</td>
<td style="text-align:right;">
19738
</td>
<td style="text-align:right;">
8799
</td>
<td style="text-align:right;">
12689
</td>
<td style="text-align:right;">
12688
</td>
<td style="text-align:right;">
9703
</td>
</tr>
<tr>
<td style="text-align:left;">
samples
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
102
</td>
<td style="text-align:right;">
89
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
57
</td>
</tr>
</tbody>
</table>
<p>As we can see, the studies have different number of genes measured and contain different number of samples. In a meta analyses, studies can have different samples sizes but genes should match between different studies. To be able to merge data sets, all data matrices should have the same annotations</p>
<p>Let us look at the row names which are already set to gene names.</p>
<pre class="r"><code>kableExtra::kable(lapply(prostate8$data, row.names) %&gt;% lapply(.,head, 5), booktabs = TRUE)</code></pre>
<table class="kable_wrapper">
<tbody>
<tr>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ACTB
</td>
</tr>
<tr>
<td style="text-align:left;">
GLRA1
</td>
</tr>
<tr>
<td style="text-align:left;">
KCNB2
</td>
</tr>
<tr>
<td style="text-align:left;">
MGAT5
</td>
</tr>
<tr>
<td style="text-align:left;">
BMP3
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
MAPK3
</td>
</tr>
<tr>
<td style="text-align:left;">
TIE1
</td>
</tr>
<tr>
<td style="text-align:left;">
CXCR5
</td>
</tr>
<tr>
<td style="text-align:left;">
DUSP1
</td>
</tr>
<tr>
<td style="text-align:left;">
MMP10
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ITGB2
</td>
</tr>
<tr>
<td style="text-align:left;">
HMBS
</td>
</tr>
<tr>
<td style="text-align:left;">
GATA6
</td>
</tr>
<tr>
<td style="text-align:left;">
ICAM5
</td>
</tr>
<tr>
<td style="text-align:left;">
RIN2
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
DDR1
</td>
</tr>
<tr>
<td style="text-align:left;">
PAX8
</td>
</tr>
<tr>
<td style="text-align:left;">
THRA
</td>
</tr>
<tr>
<td style="text-align:left;">
CCL5
</td>
</tr>
<tr>
<td style="text-align:left;">
ESRRA
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
MAPK3
</td>
</tr>
<tr>
<td style="text-align:left;">
TIE1
</td>
</tr>
<tr>
<td style="text-align:left;">
CXCR5
</td>
</tr>
<tr>
<td style="text-align:left;">
DUSP1
</td>
</tr>
<tr>
<td style="text-align:left;">
MMP10
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
DDR1
</td>
</tr>
<tr>
<td style="text-align:left;">
RFC2
</td>
</tr>
<tr>
<td style="text-align:left;">
PAX8
</td>
</tr>
<tr>
<td style="text-align:left;">
ESRRA
</td>
</tr>
<tr>
<td style="text-align:left;">
GAS6
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
DDR1
</td>
</tr>
<tr>
<td style="text-align:left;">
RFC2
</td>
</tr>
<tr>
<td style="text-align:left;">
HSPA6
</td>
</tr>
<tr>
<td style="text-align:left;">
PAX8
</td>
</tr>
<tr>
<td style="text-align:left;">
PTPN21
</td>
</tr>
</tbody>
</table>
</td>
<td>
<table>
<thead>
<tr>
<th style="text-align:left;">
x
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
ZFX
</td>
</tr>
<tr>
<td style="text-align:left;">
PIGS
</td>
</tr>
<tr>
<td style="text-align:left;">
ZPBP
</td>
</tr>
<tr>
<td style="text-align:left;">
NOTCH3
</td>
</tr>
<tr>
<td style="text-align:left;">
IGSF11
</td>
</tr>
</tbody>
</table>
</td>
</tr>
</tbody>
</table>
<p>It is clear that the gene names are set to gene symbols for all data sets and we can merge theses data set.</p>
<p>The second element of the lists corresponds to data labels where each value corresponds to disease status of the individual where the sample came from.</p>
<pre class="r"><code>prostate8$dataLabel</code></pre>
<pre><code>## $Welsh
##  [1] 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Yu
##   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [75] 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## [112] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Lapointe
##   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [38] 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
##  [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Varambally
##  [1] 0 0 0 0 0 0 1 1 1 1 1 1 1
## 
## $Singh
##   [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
##  [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
##  [75] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Wallace
##  [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## [39] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## [77] 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Nanni
##  [1] 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
## 
## $Tomlins
##  [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1
## [39] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1</code></pre>
<p>Data labels indicate “0” as control and “1” as disease groups. We can set these values to character format for convenience.</p>
<p>If everything looks good, we are good to go to the next step.</p>
<blockquote>
<p>Take a moment to explore the data in your own ways and fimilarise yourself wit it.</p>
</blockquote>
<hr />
</div>
<div id="section-data-preprocessing" class="section level2">
<h2>Data preprocessing</h2>
<div id="section-gene-matching" class="section level3">
<h3>Gene matching</h3>
<p>Usually different microarray platforms use their own probe IDs or experiments from different omics platforms can have annotations for transcripts of proteins. To perform meta-analysis, one needs to match probe/transcript IDs from different platforms to the unique official gene ID, such as ENTREZ ID or gene symbol.</p>
<p>Options for situations for microarray data include</p>
<ul>
<li>take the average value of expression values across multiple probe IDs to represent the corresponded gene symbol</li>
<li>select the probe ID with the largest interquartile range (IQR) of expression</li>
<li>Or some version of summary at the gene level depending on the omics platform.</li>
</ul>
<p>We do not cover the pre-processing steps for different omics technologies at the moment but the workshop participants are encouraged to apply the domain specific knowledge when setting up a meta analyses study.</p>
<pre class="r"><code>kableExtra::kable(MetaQC::metaOverlap(prostate8$data) %&gt;% 
                    lapply(dim) %&gt;% as.data.frame(row.names = c(&quot;genes&quot;, &quot;samples&quot;)), booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Welsh
</th>
<th style="text-align:right;">
Yu
</th>
<th style="text-align:right;">
Lapointe
</th>
<th style="text-align:right;">
Varambally
</th>
<th style="text-align:right;">
Singh
</th>
<th style="text-align:right;">
Wallace
</th>
<th style="text-align:right;">
Nanni
</th>
<th style="text-align:right;">
Tomlins
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
genes
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
</tr>
<tr>
<td style="text-align:left;">
samples
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
102
</td>
<td style="text-align:right;">
89
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
57
</td>
</tr>
</tbody>
</table>
<p>As you can see here, only 4241 genes are present in all studies and if we are to filter out this way, we lose a lot of data.</p>
<blockquote>
<p>Biologically, it is likely that most genes are either un-expressed or un-informative. In gene expression analysis to find DE genes, these genes contribute to the false discoveries, so it is desirable to filter out these genes prior to analysis. After genes are matched across studies, the unique gene symbols are available across all studies. Two sequential steps of gene filtering can be performed. In the first step, we filter out genes with very low gene expression that are identified with small average expression values across majority of studies.</p>
</blockquote>
<p>You can take a look at <a href="">preproc</a> package for some of the functions available for filtering the data and the intuition behind the methods. Here <code>Annotate(), Impute(), Filter() and Merge()</code> maybe useful for pre-processing steps of the data analyses.</p>
<pre class="r"><code>data2 &lt;- prostate8$data
data2 &lt;- preproc::Merge(data2)</code></pre>
<pre><code>## Warning in if (class(datasets[[1]]) == &quot;matrix&quot;) Merge.matrix(datasets) else
## if (class(datasets[[1]]) == : the condition has length &gt; 1 and only the first
## element will be used</code></pre>
<pre class="r"><code>data.type = rep(&quot;microarray&quot;, length(data2)) # a character vector for each study type

data2_filt &lt;- preproc::Filter(data2, del.perc = c(0.1, 0.1),data.type = data.type ) # Here we specify the percentage of genes to be filtere, and specify the type of data for each study

summary_preproc &lt;- rbind(as.data.frame(lapply(prostate8$data, dim))[1,],
                        as.data.frame(lapply(data2, dim))[1,], 
                        as.data.frame(lapply(data2_filt, dim))
                         )

rownames(summary_preproc) &lt;- c(&quot;orginal&quot;, &quot;merged&quot;, &quot;mergeFiltered&quot;, &quot;samples&quot;)

kableExtra::kable(summary_preproc, booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
Welsh
</th>
<th style="text-align:right;">
Yu
</th>
<th style="text-align:right;">
Lapointe
</th>
<th style="text-align:right;">
Varambally
</th>
<th style="text-align:right;">
Singh
</th>
<th style="text-align:right;">
Wallace
</th>
<th style="text-align:right;">
Nanni
</th>
<th style="text-align:right;">
Tomlins
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
orginal
</td>
<td style="text-align:right;">
8798
</td>
<td style="text-align:right;">
8799
</td>
<td style="text-align:right;">
13579
</td>
<td style="text-align:right;">
19738
</td>
<td style="text-align:right;">
8799
</td>
<td style="text-align:right;">
12689
</td>
<td style="text-align:right;">
12688
</td>
<td style="text-align:right;">
9703
</td>
</tr>
<tr>
<td style="text-align:left;">
merged
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
<td style="text-align:right;">
4241
</td>
</tr>
<tr>
<td style="text-align:left;">
mergeFiltered
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
<td style="text-align:right;">
3434
</td>
</tr>
<tr>
<td style="text-align:left;">
samples
</td>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
102
</td>
<td style="text-align:right;">
89
</td>
<td style="text-align:right;">
30
</td>
<td style="text-align:right;">
57
</td>
</tr>
</tbody>
</table>
<p>Note that the multiple gene expression data sets may not be very well aligned by genes, and the number of genes in each study maybe different. When we combine a large number of studies, the number of common genes may be very small, so we need to allow for genes appearing in most studies and missing in few studies etc.</p>
</div>
<div id="section-load-pathway-database" class="section level3">
<h3>Load pathway database</h3>
<p>Let us perform quality control of the data for this meta analyses. We are using the <a href="https://academic.oup.com/nar/article/40/2/e15/2408973">“MetaQC”</a> that identifies ways to objectively perform quality control for the microarray studies.</p>
<pre class="r"><code>#load(file = &quot;./data/pathways.rda&quot;)
load(file =  &quot;./data/pathwayDatabase.rda&quot;)
## Let us look into the database

pathway_names &lt;- pathwayDatabase %&gt;% names() %&gt;% stringr::str_split(pattern = &quot;_&quot;, simplify = TRUE) %&gt;% {.[,1]} %&gt;% unique()
print(pathway_names)
DList=prostate8$data
colLabel=prostate8$dataLabel
#GList=pathway[[1]]
GList=pathwayDatabase
filterGenes=TRUE
cutRatioByMean=0.2 #
cutRatioByVar=0.2

QCresult=MetaQC(DList, colLabel, GList,filterGenes,cutRatioByMean,cutRatioByVar) ## This will take some time depending on the number of studies and the type of data</code></pre>
<pre class="r"><code>kableExtra::kable(QCresult$scoreTable, booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
IQC
</th>
<th style="text-align:right;">
EQC
</th>
<th style="text-align:right;">
AQCg
</th>
<th style="text-align:right;">
AQCp
</th>
<th style="text-align:right;">
CQCg
</th>
<th style="text-align:right;">
CQCp
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Welsh
</td>
<td style="text-align:right;">
3.9754526
</td>
<td style="text-align:right;">
4.0671059
</td>
<td style="text-align:right;">
33.3693516
</td>
<td style="text-align:right;">
4.770316
</td>
<td style="text-align:right;">
188.3863094
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Yu
</td>
<td style="text-align:right;">
9.6777279
</td>
<td style="text-align:right;">
8.7893559
</td>
<td style="text-align:right;">
29.8963097
</td>
<td style="text-align:right;">
12.931269
</td>
<td style="text-align:right;">
172.8240650
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Lapointe
</td>
<td style="text-align:right;">
3.7134470
</td>
<td style="text-align:right;">
4.4921021
</td>
<td style="text-align:right;">
25.0966984
</td>
<td style="text-align:right;">
16.416134
</td>
<td style="text-align:right;">
100.7034534
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Varambally
</td>
<td style="text-align:right;">
3.7134470
</td>
<td style="text-align:right;">
3.5224557
</td>
<td style="text-align:right;">
4.5085131
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
26.6559834
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Singh
</td>
<td style="text-align:right;">
1.7315604
</td>
<td style="text-align:right;">
1.8478981
</td>
<td style="text-align:right;">
16.3348915
</td>
<td style="text-align:right;">
66.488268
</td>
<td style="text-align:right;">
48.2152305
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Wallace
</td>
<td style="text-align:right;">
8.1666052
</td>
<td style="text-align:right;">
8.8880394
</td>
<td style="text-align:right;">
0.0364897
</td>
<td style="text-align:right;">
25.144610
</td>
<td style="text-align:right;">
0.0060950
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Nanni
</td>
<td style="text-align:right;">
1.0232669
</td>
<td style="text-align:right;">
1.9239147
</td>
<td style="text-align:right;">
0.0000000
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
0.6822836
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Tomlins
</td>
<td style="text-align:right;">
0.4933381
</td>
<td style="text-align:right;">
0.5033783
</td>
<td style="text-align:right;">
1.9044123
</td>
<td style="text-align:right;">
10.330613
</td>
<td style="text-align:right;">
12.8201866
</td>
<td style="text-align:right;">
410
</td>
</tr>
</tbody>
</table>
<p>IQC, EQC, AQCg, AQCp, CQCg, CQCp</p>
<blockquote>
<p>Internal quality control index:small IQC indicated that the study had heterogeneous coexpression structure with other studies and was considered a candidate problematic study that should be excluded from meta-analysis</p>
<p>the external quality control (EQC): small <em>EQC</em> indicated that the study had low association with pathway in terms of gene pairwise correlation structure and maybe considered a candidate problematic study.</p>
<p>accuracy quality control (<em>AQC</em>) and a consistency quality control (<em>CQC</em>).</p>
<p>Large <em>AQCg</em> measure for a given study indicate that DE genes produced by study were reproducible compared to DE genes detected by meta-analysis excluding study</p>
<p>Having a large CQCg measure for a given study indicated that DE evidence produced by study was consistent with DE evidence generated by meta-analysis excluding study. Please read the original <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3898528/#B14">MetQC publication</a> for a better understanding of the measures.</p>
</blockquote>
<p>We can now look at the two dimensional biplot based on PCA analyses and get an idea of any studies that may not be of great quality.</p>
<pre class="r"><code>MetaQC::plotMetaQC(QCresult$scoreTable)</code></pre>
<p><img src="mata-analyses-in-r_files/figure-html/qc-plot-1.png" width="624" /></p>
<p>Although the <code>plotMetaQC()</code> function is a useful tool for visualization, it does not necessarily allow automatic inclusion/exclusion criteria.</p>
<p>Note that our visualization and summarization tools were not meant for an automated recommendation for inclusion/exclusion decision. In the examples we explored, there were roughly three categories in the QC results: definite exclusion cases with poor quality, definite inclusion cases with good quality and borderline cases. </p>
<p>Here Nanni and Tomlins are the two studies that maybe of low quality and should perhaps be excluded from the analyses.</p>
<pre class="r"><code>filterGenes=TRUE
cutRatioByMean=0.2 #
cutRatioByVar=0.2
to_remove &lt;- c(&quot;Nanni&quot;, &quot;Tomlins&quot;)
prostate6 &lt;- list(data = within(prostate8$data, rm(Nanni, Tomlins)), dataLabel=within(prostate8$dataLabel, rm(Nanni, Tomlins)))
prostate_fil &lt;- list(data = MetaQC::metaOverlap(prostate6$data), dataLabel=prostate6$data)
prostate6$data &lt;-MetaQC::metaFilterData(prostate_fil$data,cutRatioByVar = cutRatioByVar, cutRatioByMean = cutRatioByMean)
lapply(prostate6, names)</code></pre>
<pre><code>## $data
## [1] &quot;Welsh&quot;      &quot;Yu&quot;         &quot;Lapointe&quot;   &quot;Varambally&quot; &quot;Singh&quot;     
## [6] &quot;Wallace&quot;   
## 
## $dataLabel
## [1] &quot;Welsh&quot;      &quot;Yu&quot;         &quot;Lapointe&quot;   &quot;Varambally&quot; &quot;Singh&quot;     
## [6] &quot;Wallace&quot;</code></pre>
<pre class="r"><code>kableExtra::kable(as.data.frame(lapply(prostate6$data, dim)), booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:right;">
Welsh
</th>
<th style="text-align:right;">
Yu
</th>
<th style="text-align:right;">
Lapointe
</th>
<th style="text-align:right;">
Varambally
</th>
<th style="text-align:right;">
Singh
</th>
<th style="text-align:right;">
Wallace
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
4441
</td>
<td style="text-align:right;">
4441
</td>
<td style="text-align:right;">
4441
</td>
<td style="text-align:right;">
4441
</td>
<td style="text-align:right;">
4441
</td>
<td style="text-align:right;">
4441
</td>
</tr>
<tr>
<td style="text-align:right;">
34
</td>
<td style="text-align:right;">
146
</td>
<td style="text-align:right;">
103
</td>
<td style="text-align:right;">
13
</td>
<td style="text-align:right;">
102
</td>
<td style="text-align:right;">
89
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>DList=prostate6$data
colLabel=prostate6$dataLabel
#GList=pathway[[1]]
GList=pathwayDatabase
filterGenes=TRUE
cutRatioByMean=0.2 #
cutRatioByVar=0.2
QCresult2=MetaQC(DList, colLabel, GList,filterGenes,cutRatioByMean,cutRatioByVar)</code></pre>
<pre class="r"><code>kableExtra::kable(QCresult2$scoreTable, booktabs = TRUE)</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
IQC
</th>
<th style="text-align:right;">
EQC
</th>
<th style="text-align:right;">
AQCg
</th>
<th style="text-align:right;">
AQCp
</th>
<th style="text-align:right;">
CQCg
</th>
<th style="text-align:right;">
CQCp
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Welsh
</td>
<td style="text-align:right;">
3.7730194
</td>
<td style="text-align:right;">
2.784275
</td>
<td style="text-align:right;">
71.8151852
</td>
<td style="text-align:right;">
6.006001
</td>
<td style="text-align:right;">
278.276330
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Yu
</td>
<td style="text-align:right;">
7.6904488
</td>
<td style="text-align:right;">
8.582181
</td>
<td style="text-align:right;">
58.3058931
</td>
<td style="text-align:right;">
4.909503
</td>
<td style="text-align:right;">
252.078158
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Lapointe
</td>
<td style="text-align:right;">
1.6703062
</td>
<td style="text-align:right;">
1.281574
</td>
<td style="text-align:right;">
39.7832310
</td>
<td style="text-align:right;">
4.776324
</td>
<td style="text-align:right;">
134.483776
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Varambally
</td>
<td style="text-align:right;">
3.0322791
</td>
<td style="text-align:right;">
3.470636
</td>
<td style="text-align:right;">
8.9101248
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
35.007095
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Singh
</td>
<td style="text-align:right;">
0.7670506
</td>
<td style="text-align:right;">
1.385652
</td>
<td style="text-align:right;">
24.1760513
</td>
<td style="text-align:right;">
9.603674
</td>
<td style="text-align:right;">
69.551325
</td>
<td style="text-align:right;">
410
</td>
</tr>
<tr>
<td style="text-align:left;">
Wallace
</td>
<td style="text-align:right;">
6.4044351
</td>
<td style="text-align:right;">
5.736912
</td>
<td style="text-align:right;">
0.0038734
</td>
<td style="text-align:right;">
0.000000
</td>
<td style="text-align:right;">
0.000635
</td>
<td style="text-align:right;">
410
</td>
</tr>
</tbody>
</table>
<pre class="r"><code>plotMetaQC(scoreTable = QCresult2$scoreTable)</code></pre>
<p><img src="mata-analyses-in-r_files/figure-html/qc-plot-2-1.png" width="624" /></p>
</div>
</div>
<div id="section-meta-differential-expression-analyses" class="section level2">
<h2>Meta Differential Expression analyses</h2>
<p>We have listed the methods for differential expression meta analyses in the introduction section that are implemented in the MetaDE package. Here we will try the best performing methods including AW Fischer, REM and rOP. A detailed comparison of relative performances of each of the statistical methods is descibed <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3898528/#B14">Lun-Ching etal</a>.</p>
<pre class="r"><code>data &lt;- prostate6$data # Extract the expression matrices
clin.data &lt;- prostate6$dataLabel # extract lables for each sample

K &lt;- length(data)

clin.data &lt;- lapply(clin.data, function(x) {data.frame(x)} )

for (k in 1:length(clin.data)){
 colnames(clin.data[[k]]) &lt;- &quot;label&quot;
 clin.data[[k]] &lt;- (ifelse(clin.data[[k]]==0, &quot;Control&quot;, &quot;Cancer&quot;))

}

#clin.data &lt;- lapply(clin.data, function(x) {data.frame(x)} )
select.group &lt;- c(&quot;Control&quot;, &quot;Cancer&quot;)
ref.level &lt;- &quot;Control&quot;
data.type &lt;- &quot;continuous&quot;
ind.method &lt;- rep(&#39;limma&#39;,length(data))
resp.type &lt;- &quot;twoclass&quot;
paired &lt;- rep(FALSE,length(data))


meta.method &lt;- &quot;Fisher&quot;


meta.res &lt;- MetaDE(data=data,clin.data = clin.data,
                    data.type=data.type,resp.type = resp.type,
                    response=&#39;label&#39;,
                    ind.method=ind.method, meta.method=meta.method,
                    select.group = select.group, ref.level=ref.level,
                    paired=paired,tail=&#39;abs&#39;,parametric=TRUE)</code></pre>
<pre><code>## Please make sure the following is correct:
## *You input 6 studies
## *You selected limma limma limma limma limma limma for your 6 studies respectively
## * Fisher was chosen to combine the 6 studies,respectively
## dataset 1 is done
## dataset 2 is done
## dataset 3 is done
## dataset 4 is done
## dataset 5 is done
## dataset 6 is done
## Parametric method was used instead of permutation</code></pre>
<pre class="r"><code>head(MetaDE::summary.meta(meta.res, resp.type = resp.type, meta.method = meta.method))</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["ind.stat.Welsh"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["ind.stat.Yu"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["ind.stat.Lapointe"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["ind.stat.Varambally"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["ind.stat.Singh"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["ind.stat.Wallace"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ind.p.Welsh"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["ind.p.Yu"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["ind.p.Lapointe"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["ind.p.Varambally"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["ind.p.Singh"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["ind.p.Wallace"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["stat"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["pval"],"name":[14],"type":["dbl"],"align":["right"]},{"label":["FDR"],"name":[15],"type":["dbl"],"align":["right"]}],"data":[{"1":"-0.01447604","2":"0.37943628","3":"0.28513849","4":"0.1086048","5":"-0.106957489","6":"-0.330154800","7":"0.9787124856","8":"6.535083e-03","9":"2.766436e-01","10":"0.82103641","11":"0.0352376882","12":"5.156390e-02","13":"25.68975","14":"1.187213e-02","15":"1.629300e-02","_rn_":"GPR12"},{"1":"0.20673665","2":"0.23162454","3":"0.77668332","4":"0.1880854","5":"1.026749249","6":"-0.476614732","7":"0.0158808041","8":"2.471170e-03","9":"3.604894e-10","10":"0.13938329","11":"0.0005260721","12":"6.423977e-06","13":"106.73068","14":"2.646549e-17","15":"5.657944e-16","_rn_":"RPS19"},{"1":"-0.11916266","2":"0.11168920","3":"0.36585563","4":"0.3870752","5":"0.005348824","6":"0.282557189","7":"0.4283140887","8":"4.380445e-01","9":"8.016986e-02","10":"0.05144596","11":"0.9190869107","12":"2.078376e-02","13":"22.24424","14":"3.487187e-02","15":"4.358622e-02","_rn_":"GALNT2"},{"1":"-0.28057451","2":"-0.03188107","3":"-0.22775570","4":"-0.1435580","5":"-0.185613924","6":"-0.024422808","7":"0.5518416771","8":"7.150206e-01","9":"3.189142e-03","10":"0.26616084","11":"0.0512139773","12":"7.256761e-01","13":"22.58798","14":"3.143462e-02","15":"3.959193e-02","_rn_":"TTLL5"},{"1":"-0.57566042","2":"0.25101054","3":"0.04482983","4":"0.2506372","5":"-0.018258782","6":"-0.008988302","7":"0.2972461810","8":"7.932261e-02","9":"8.619937e-01","10":"0.24834054","11":"0.6977351235","12":"9.462380e-01","13":"11.40813","14":"4.943030e-01","15":"5.069745e-01","_rn_":"ERN1"},{"1":"0.31611140","2":"0.30025227","3":"0.71148387","4":"0.1270682","5":"1.181183549","6":"-0.147041998","7":"0.0001842776","8":"7.488635e-07","9":"1.080361e-05","10":"0.26787502","11":"0.0023606587","12":"4.036376e-02","13":"89.43056","14":"6.363482e-14","15":"7.436901e-13","_rn_":"RPL28"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<div id="section-plot-results-in-a-heatmap." class="section level3">
<h3>Plot results in a heatmap.</h3>
<pre class="r"><code>fdr.cut &lt;- 1e-7
pdf(&#39;heatmap.test.pdf&#39;)
heatmap.sig.genes(meta.res, meta.method=meta.method,
                    fdr.cut=fdr.cut,color=&quot;GR&quot;)
dev.off()</code></pre>
<div class="figure">
<img src="images/heatmap-sig-genes.png" style="height:150.0%" alt="" />
<p class="caption">heatmap</p>
</div>
</div>
<div id="section-pathway-analyses-based-on-meta-analyses" class="section level3">
<h3>Pathway analyses based on meta analyses</h3>
<p>At this stage, one can do pathway analyses directly on the genes that appear to be differentially regulated based on the meta analyses performed above.</p>
<pre class="r"><code>meta.p  &lt;- meta.res$meta.analysis$pval
ks.result &lt;- PathAnalysis(meta.p = meta.p, enrichment = &quot;KS&quot;)
fisher.result &lt;- PathAnalysis(meta.p = meta.p, enrichment = &quot;Fisher&#39;s exact&quot;)

path.res &lt;- MetaDE::PathAnalysis(meta.p ,pathway = pathwayDatabase, p.cut = 0.05,
                    enrichment = &quot;Fisher&#39;s exact&quot;, DEgene.number = 400, size.min = 10, size.max = 500)</code></pre>
<blockquote>
<p>The pathway analyses performed here is based on the p-values obtained after the joint meta analyses of all the studies in your data. Some time individual study level pathway analyses is reported and we are interested in combining pathway level summary statistics. We will explore that in our next session where we will perform the analyses using <strong>“MetaPath”</strong> package.</p>
</blockquote>
</div>
</div>
<div id="section-exercise" class="section level2">
<h2>Exercise</h2>
<ul>
<li><p>Perform DE meta analyses using any two methods among “AW-Fisher”, “FEM,”REM and “minMCC” and compare the number of differentially expressed genes detected at FDR &lt;0.05. Take a look at the /Users/cob-aaf/Library/Application Support/renv/cache/v5/R-4.0/x86_64-apple-darwin17.0/MetaDE/2.2.3/8598944739859703e37b4623921341ac/MetaDE/help/MetaDE function to change the parameters.</p></li>
<li><p>Perform pathway analyses based one of the methods you tried and see whether any differences at pathway level were detected.</p></li>
</ul>
<div class="tutorial-exercise" data-label="change-de-method" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>meta.method &lt;- &quot;AW&quot;
meta.res &lt;- MetaDE(data=data,clin.data = clin.data,
                    data.type=data.type,resp.type = resp.type,
                    response=&#39;label&#39;,covariate = NULL,
                    ind.method=ind.method, meta.method=meta.method,
                    select.group = select.group, ref.level=ref.level,
                    paired=paired, rth=NULL,
                    REM.type=NULL,tail=&#39;abs&#39;,parametric=TRUE)</code></pre>
<script type="application/json" data-opts-chunk="1">{"fig.width":6.5,"fig.height":4,"fig.retina":2,"fig.align":"default","fig.keep":"high","fig.show":"asis","out.width":624,"warning":true,"error":false,"message":true,"exercise.df_print":"paged","exercise.checker":"NULL"}</script>
</div>
<div class="tutorial-exercise-support" data-label="change-de-method-hint-1" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>meta.method &lt;- &quot;FEM&quot;
meta.res &lt;- MetaDE(data=data,clin.data = clin.data,
                    data.type=data.type,resp.type = resp.type,
                    response=&#39;label&#39;,
                    ind.method=ind.method, meta.method=meta.method,
                    select.group = select.group, ref.level=ref.level,
                    paired=paired, tail=&#39;abs&#39;)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="change-de-method-hint-2" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>meta.method &lt;- &quot;REM&quot;
REM.type &lt;- &quot;HO&quot;
meta.res &lt;- MetaDE(data=data,clin.data = clin.data,
                    data.type=data.type,resp.type = resp.type,
                    response=&#39;label&#39;,
                    ind.method=ind.method, meta.method=meta.method,
                    select.group = select.group, ref.level=ref.level,
                    paired=paired,
                    REM.type=REM.type,tail=&#39;abs&#39;)</code></pre>
</div>
<div class="tutorial-exercise-support" data-label="change-de-method-hint-3" data-caption="Code" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0">
<pre class="text"><code>meta.method &lt;- &#39;minMCC&#39;
meta.res &lt;- MetaDE(data=data,clin.data = clin.data,
                    data.type=data.type,resp.type = resp.type,
                    response=&#39;label&#39;,
                    ind.method=ind.method, meta.method=meta.method,
                    select.group = select.group, ref.level=ref.level,
                    paired=paired,tail=&#39;abs&#39;,parametric=FALSE,nperm=100)</code></pre>
</div>
</div>
<div id="section-pathway-analyses" class="section level2">
<h2>Pathway analyses</h2>
<p>Pathway analysis (a.k.a. gene set analysis) is a statistical tool to infer correlation of differential expression evidence in the data with pathway knowledge from established databases. The idea behind pathway analysis is to determine if there is enrichment in the detected DE genes based on an <em>a priori</em> defined biological category. Such a category might come from one or multiple databases such as Gene Ontology (GO; <a href="http://www.geneontology.org/">www.geneontology.org</a>), the Kyoto Encyclopedia of Genes and Genomes (KEGG; <a href="http://www.genome.jp/kegg/" class="uri">http://www.genome.jp/kegg/</a>), Biocarta Pathways (<a href="http://www.biocarta.com/" class="uri">http://www.biocarta.com/</a>) and the comprehensive Molecular Signatures Database (MSigDB; <a href="http://www.broadinstitute.org/gsea/msigdb/" class="uri">http://www.broadinstitute.org/gsea/msigdb/</a>). For the majority of recent microarray meta-analysis applications, pathway analysis has been a standard follow-up to identify pathways associated with detected DE genes <a href="https://pubmed.ncbi.nlm.nih.gov/17974971/">e.g.</a> and many others]. The result provides more insightful biological interpretation and it has been reported that pathway analysis results are usually more consistent and reproducible across studies than DE gene detection. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2865865/">Shen and Tseng</a> developed a systematic framework of Meta-Analysis for Pathway Enrichment (MAPE) by combining information at gene level, at pathway level and a hybrid of the two.</p>
<div id="section-step1" class="section level4">
<h4>Step1</h4>
<pre class="r"><code>load(&quot;./data/pathways.rda&quot;)

CPI_result = MAPE2.0(arraydata = data , clinical.data = clin.data,
                    label = &quot;label&quot;,pmtx = NULL,pathway = c(Biocarta.genesets, GOBP.genesets, 
        GOCC.genesets, GOMF.genesets, KEGG.genesets, Reactome.genesets), data.type =&quot;discrete&quot;, 
        resp.type = &quot;twoclass&quot;,method = &quot;CPI&quot;, ind.method = rep(&quot;limma&quot;,length(data)),
        paired =rep(FALSE,length(data)),select.group=select.group, ref.level=ref.level ,
        tail=&quot;abs&quot;, enrichment = &quot;Fisher&#39;s exact&quot;, DEgene.number = 400,stat = &quot;AW Fisher&quot;)</code></pre>
</div>
<div id="section-step-2" class="section level3">
<h3>Step 2</h3>
<pre class="r"><code>set.seed(15213)
CPI.kappa_result = MAPE.Kappa(summary = CPI_result$summary,pathway = CPI_result$pathway,
                              max_k = 8, q_cutoff = 0.0005,software = CPI_result$method, output_dir = &quot;./results&quot;)</code></pre>
<p><img src="images/consensus012.png" /></p>
<pre class="r"><code>#data(hashtb)
data(&quot;./data/hashtb.rda&quot;)
#xx &lt;- data(hashtb)
set.seed(111222333)
MAPE.Clustering.results &lt;- MAPE.Clustering(summary=CPI_result$summary,Num_Clusters = 10,
                                kappa.result = CPI.kappa_result$kappa,sil_cut=0.1,
                                Num_of_gene_lists=CPI_result$Num_of_gene_lists,genelist =CPI_result$genelist,
                                pathway=CPI_result$pathway, enrichment=CPI_result$enrichment,
                                method=CPI.kappa_result$method,software=CPI_result$method,
                                n.text.permute = 10000, output_dir = &quot;./results&quot;  )</code></pre>
<div class="figure">
<img src="images/consensus013.png" alt="" />
<p class="caption">cluster membershio scoring</p>
</div>
<div class="figure">
<img src="images/Heatmap_clusters_all.png" style="height:120.0%" alt="" />
<p class="caption">consensus clustering</p>
</div>
<hr />
</div>
</div>
<div id="section-network-meta-analyses-using-metadcn" class="section level2">
<h2>Network Meta analyses using MetaDCN</h2>
<p>Co-expression analysis and network analysis of microarray data are used to investigate potential transcriptional co-regulation and gene interactions. Network analyses typically work with the gene–gene co-expression matrix, which represents the correlation between each pair of genes in the study. A crucial assumption is that the magnitude of the co-expression between any pair of genes is associated with a greater likelihood that the two genes interact. Thus, networks of interactions between genes are inferred from the co-expression matrix.</p>
<p>Here we will use <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6041767/">MetaDCN</a> to extract network modules in disease vs control conditions.</p>
<div class="figure">
<img src="images/metaDCN.jpg" alt="" />
<p class="caption">MetaDCN</p>
</div>
<pre class="r"><code>data(&quot;./data/pathwayDatabase.rda&quot;)
#data(example)
# Generate correction and adjacency matrices for data and permutation
# GeneNet returns a list of information which will be used for SearchBM and MetaDCN function, and several RData files stored in folder path.
# AdjacencyMatrices.RData is a list of adjacency matrices for case and control in each study in the order of case studies and control studies.
# 
# CorrelationMatrices.RData is a list of correlation matrices for case and control in each study.
# 
# AdjacencyMatricesPermutationP.RData is a list of correlation matrices for case and control in each study in permutation P.
GeneNetRes.2  &lt;- GeneNet(data , clin.data, caseName=&quot;Cancer&quot;, controlName=&quot;Control&quot;, meanFilter=0.8, SDFilter=0.8, edgeCutoff=0.1, permutationTimes=4, CPUNumbers=4,  pathwayDatabase=c(Biocarta.genesets,GOBP.genesets,GOCC.genesets,GOMF.genesets,
                    KEGG.genesets,Reactome.genesets), silent=FALSE, folder = &quot;./MetaDCN&quot;)</code></pre>
<pre class="r"><code># This function will search for basic modules differentially co-expressed between case and control
# SearchBM will return a list and several Rdata, csv and png files saved in the folder path specified in GeneNet inputs.List of basic module information:

# w1    
# w1 weight with the most basic modules detected
# 
# BMInCase  
# data matrix listing the information of basic modules higher correlated in case
# 
# BMInControl   
# data matrix listing the information of basic modules higher correlated in control
# 
# permutation_energy_direction_p.Rdata is a list of energies of basic modules from permutation p.
# 
# basic_modules_summary_direction_weight_w.csv is a summary of basic modules detected using weight w in forward/backward search.
# 
# threshold_direction.csv is a table listing number of basic modules detected under different FDRs in forward/backward search.
# 
# figure_basic_module_c_repeat_r_direction_weight_w.png is a plot of basic module from component c repeat r using weight w in forward/backward search.

SearchBMRes &lt;- SearchBM(GeneNetRes.2, MCSteps=500, jaccardCutoff=0.8, repeatTimes=3, outputFigure=TRUE, silent=FALSE)</code></pre>
<pre class="r"><code>#This function will assemble basic modules detected from SearchBM into supermodules.
# w1:w1 used
# 
# BMInCaseSig: Summary of basic modules higher correlated in Case controling FDR
# 
# BMInControlSig    
# Summary of basic modules higher correlated in Control controling FDR
# 
# Supermodule: Summary of supermodules
#A number chosen from (100, 200, ..., 700) to specify the weight1 used in objective function (optional). If not specified, w1 from SearchBM function will be used (recommended).


MetaDNCRes &lt;- MetaDCN(GeneNetRes.2, SearchBMRes, FDRCutoff=0.05, w1=NULL, silent=FALSE)</code></pre>
<p><img src="images/Basic_module_component_1_repeat_2_weight_700_forward.png" /></p>
<p>You get the modules, their pathway annotations and cytoscape files in the results directory.</p>

<script type="application/shiny-prerendered" data-context="server-start">
library(learnr)
knitr::opts_chunk$set(echo = TRUE)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
session$onSessionEnded(function() {
        learnr:::session_stop_event(session)
      })
</script>
 
<script type="application/shiny-prerendered" data-context="data">
load(file = "./data/prostate8.rda")
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-change-de-method-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-change-de-method-code-editor`)), session)
output$`tutorial-exercise-change-de-method-output` <- renderUI({
  `tutorial-exercise-change-de-method-result`()
})
</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["header-attrs"]},{"type":"character","attributes":{},"value":["2.7"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pandoc"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["header-attrs.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/cerulean.min.css"]},{"type":"character","attributes":{},"value":["<style>h1 {font-size: 34px;}\n       h1.title {font-size: 38px;}\n       h2 {font-size: 30px;}\n       h3 {font-size: 24px;}\n       h4 {font-size: 18px;}\n       h5 {font-size: 16px;}\n       h6 {font-size: 12px;}\n       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}\n       pre:not([class]) { background-color: white }<\/style>"]},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css","rstudio-theme.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["5.1.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.css","css/v4-shims.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["2.7"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["4.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.2.6"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["1.5.15"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.1"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150]}},"value":[{"type":"character","attributes":{},"value":["abind","annotate","AnnotationDbi","assertthat","base","Biobase","BiocGenerics","BiocParallel","bit","bit64","bitops","blob","bslib","cachem","caTools","class","cluster","colorspace","combinat","compiler","ConsensusClusterPlus","crayon","curl","datasets","DBI","debugme","DelayedArray","DESeq2","digest","DMwR","dplyr","edgeR","ellipsis","evaluate","fansi","fastmap","fs","genefilter","geneplotter","generics","GenomeInfoDb","GenomeInfoDbData","GenomicRanges","ggplot2","glue","gplots","graph","graphics","grDevices","grid","GSA","GSEABase","gtable","gtools","highr","htmltools","htmlwidgets","httpuv","httr","igraph","impute","IRanges","irr","jquerylib","jsonlite","kableExtra","KernSmooth","knitr","later","lattice","learnr","lifecycle","limma","locfit","lpSolve","magrittr","markdown","MASS","Matrix","MatrixGenerics","matrixStats","memoise","MetaDCN","MetaDE","MetaPath","MetaQC","methods","mime","multtest","munsell","openxlsx","parallel","pillar","pkgconfig","plyr","preproc","promises","purrr","quantmod","R6","RColorBrewer","Rcpp","RCurl","Rgraphviz","rlang","rmarkdown","ROCR","rpart","rprojroot","RSQLite","rstudioapi","rvest","S4Vectors","samr","sass","scales","shape","shiny","shinyFiles","snow","snowfall","splines","stats","stats4","stringi","stringr","SummarizedExperiment","survival","svglite","systemfonts","tibble","tidyselect","tools","TTR","utf8","utils","vctrs","viridisLite","webshot","withr","xfun","XML","xml2","xtable","xts","XVector","yaml","zip","zlibbioc","zoo"]},{"type":"character","attributes":{},"value":["1.4-5","1.68.0","1.52.0","0.2.1","4.0.3","2.50.0","0.36.0","1.24.1","4.0.4","4.0.5","1.0-6","1.2.1","0.2.4","1.0.4","1.18.2","7.3-18","2.1.1","2.0-0","0.0-8","4.0.3","1.54.0","1.4.1","4.3","4.0.3","1.1.1","1.1.0","0.16.3","1.30.1","0.6.27","0.4.1","1.0.5","3.32.1","0.3.1","0.14","0.4.2","1.1.0","1.5.0","1.72.1","1.68.0","0.1.0","1.26.7","1.2.4","1.42.0","3.3.3","1.4.2","3.1.1","1.68.0","4.0.3","4.0.3","4.0.3","1.03.1","1.52.1","0.3.0","3.8.2","0.8","0.5.1.1","1.5.3","1.5.5","1.4.2","1.2.6","1.64.0","2.24.1","0.84.1","0.1.3","1.7.2","1.3.4","2.23-18","1.31","1.1.0.1","0.20-41","0.10.1","1.0.0","3.46.0","1.5-9.4","5.6.15","2.0.1","1.1","7.3-53.1","1.3-2","1.2.1","0.58.0","2.0.0","0.0-1","2.2.3","2.0","1.0","4.0.3","0.10","2.46.0","0.5.0","4.2.3","4.0.3","1.5.1","2.0.3","1.8.6","1.2-4","1.2.0.1","0.3.4","0.4.18","2.5.0","1.1-2","1.0.6","1.98-1.3","2.34.0","0.4.10","2.7","1.0-11","4.1-15","2.0.2","2.2.5","0.13","1.0.0","0.28.1","3.0","0.3.1","1.1.1","1.4.5","1.6.0","0.9.0","0.4-3","1.84-6.1","4.0.3","4.0.3","4.0.3","1.5.3","1.4.0","1.20.0","3.2-10","2.0.0","1.0.1","3.1.0","1.1.0","4.0.3","0.24.2","1.2.1","4.0.3","0.3.7","0.3.0","0.5.2","2.4.1","0.22","3.99-0.6","1.3.2","1.8-4","0.12.1","0.30.0","2.2.1","2.1.1","1.36.0","1.8-9"]}]}]}
</script>
<!--/html_preserve-->
</div>

</div> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h2 class="title toc-ignore" style="display:none;">Tutorial</h2>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</div> <!-- bandContent page -->
</div> <!-- pageContent band -->




<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>


</body>

</html>
