<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>Omics Data management</title>
    <meta charset="utf-8" />
    <meta name="author" content="Ashfaq Ali" />
    <link href="libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
    <link href="libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="libs/font-awesome-5.1.0/fonts/fontawesome-webfont.ttf"/>
    <link rel="stylesheet" href="assets/presentation.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Omics Data management
## Some useful tools and strategies
### Ashfaq Ali
### NBIS, SciLife lab, Lund University, Sweden
### 2019/08/14 (updated: 2019-09-08)

---

exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;





---
class: inverse, left, middle
## Why have a data management strategy?

**It leads us to follow basic scietific pronciples**

 -- *Repoducibility*
  
 -- *Avoid redunduancy*   
 
 -- *Productivity*  
 
 -- *Collaboration*
  
 -- *Creativity* 
---
class: inverse, left, middle
## Outline: Data Management Lecture 


--**Background, Basics and resources**
 - Componenets of data for reporduciple research
 - Basic data types in R (Vectors, matrices, data frames and Lists). [Tutorial](link)
 - RMarkdown, Rdata, RDS objects to store and access data on disk 

-- **Bioconductor packages and objects useful for omics data analyses**

 - S3 and S4 class objects and methods 
 - IRAnges, GRanges, GenomicsRanges, BSGenomes, , BSGenome, BioMArt
 - ExpressionSet, SummarisedExperiments, MultiAssayExperiments

-- **Data structures and manipulation in Python (Rui Benfeitas)** 



.footnote[
*Complementary tutorial is available with some practical examples and R codes.* 

*Here we donot cover pre-processing of data from different platforms and assume that we will work with the processed data.*  
]

---
class: inverse, left, middle
## Componenets of data for reporduciple research

 - **FAIR data guidelins**
Findability, Accessibility, Interoperability, and Reusability [FAIR](https://www.nature.com/articles/sdata201618)

 - **Raw Data**
 Raw data represent data from the facility where the data was obtaned from. Raw reads, spectra, images, counts etc. This depends on at what stage you boarded the ship.  
 
 - **Tiday Data**
Organized and it's shareable and ready to be used.   
More on it later.

 - **Code Book**
 This would represent things like a) What 0,1 codes represent b) Measurement scales (cm, ng/ml etc).  
 R has an excellent way of storing this information in the form of factors (tutorial)  
 
 - **Scripts, Codes, Recipie, Github**
 An easy way to keep your analyses 'recipe' is by documentin the whole analyses or set of analyses in a script.   
 A text/lig file for steps in ***GUI's*** (Cytoscape, MZmine etc).
 [NBIS Course](https://nbis-reproducible-research.readthedocs.io/en/latest/)

 - **Sensitive data and GDPR compliance**
 Be aware of the rules about person ***senstive data***.
 Read more about at these pages [NBIS Website](https://nbis.se/support/human-data.html), [GDPR](https://gdpr-info.eu), [Data Protection](https://ec.europa.eu/info/law/law-topic/data-protection/reform_en).


---
class: inverse, left, middle
## Basic data types in R

 - **Vectors**: Represent a set of single type of data such as string,    Numerics an dintegers  
 
  - ***Fctors*** are special vectors in R that allow dummy  
  coding of elements. *Very helpful but source of confuion at times.*

 - **Matrices** hold numerical data in two dimensions. 
 
 - **Arrays** Multodimensional 
 
 - **Data Frame**  can hold any kind of data.   
 
 - **Lists** are very useful data structure for Bioindormatics
  Hold any type of data with any dimention.
  
 - **S3 and S4 system** are very central to R programming: 
allowing to create objects to represent any data types,
 access and manupulate data using ***accessor*** functions. More later
  


---
class: inverse, left, middle
## A well known example linear model


```r
mat&lt;- matrix(rnorm(100), 25,4)
dat&lt;- as.data.frame(cbind(rowSums( c(0.01, 0.02, 0.03, 0.04)* mat),mat))
colnames(dat) &lt;- c("y","x1","x2","x3","x4")
fit&lt;- lm(data = as.data.frame(dat), formula = y~x1+x2+x3+x4)
print(class(fit) ); print(typeof(fit))
```

```
## [1] "lm"
## [1] "list"
```
All functions applicable to lists are useable but it is reccomened to use class specific functions. Try `methods(summary)` on your computer. or try `methods(class="GRanges")`

```r
summary(fit)
```

```
## 
## Call:
## lm(formula = y ~ x1 + x2 + x3 + x4, data = as.data.frame(dat))
## 
## Residuals:
##       Min        1Q    Median        3Q       Max 
## -0.037667 -0.010458 -0.000510  0.009894  0.036828 
## 
## Coefficients:
##              Estimate Std. Error t value Pr(&gt;|t|)    
## (Intercept) -0.002494   0.004313  -0.578  0.56960    
## x1           0.024329   0.003460   7.031 8.07e-07 ***
## x2           0.028836   0.005069   5.689 1.44e-05 ***
## x3           0.018645   0.005585   3.339  0.00327 ** 
## x4           0.031598   0.004119   7.671 2.21e-07 ***
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.02053 on 20 degrees of freedom
## Multiple R-squared:  0.9157,	Adjusted R-squared:  0.8989 
## F-statistic: 54.34 on 4 and 20 DF,  p-value: 1.831e-10
```


---
class: inverse, left, middle
## Bioconductor packages and objects useful for omics data analyses



**IRanges class to represent data that occurs in a sequence.**

An IRanges **constructor** can be used to create and object  
of ***IRanges class***


```r
ir &lt;- IRanges(start=c(8, 12, 17, 21), width=c(4, 5, 4, 10),   
              names = c("Morning", "Afternoon", "Evening", "Night"))
print(ir[1:2,])
```

```
## IRanges object with 2 ranges and 0 metadata columns:
##                 start       end     width
##             &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##     Morning         8        11         4
##   Afternoon        12        16         5
```

One can apply class specific methods to objects.

---
class: inverse, left, middle
## GenomicsRAnges

Extention of `IRanges`


```r
library(GenomicRanges)
gr &lt;- GRanges( seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
gr [1:4,]
```

```
## GRanges object with 4 ranges and 2 metadata columns:
##     seqnames    ranges strand |     score                GC
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;         &lt;numeric&gt;
##   a     chr1   101-111      - |         1                 1
##   b     chr2   102-112      + |         2 0.888888888888889
##   c     chr2   103-113      + |         3 0.777777777777778
##   d     chr2   104-114      * |         4 0.666666666666667
##   -------
##   seqinfo: 3 sequences from an unspecified genome; no seqlengths
```




---

## SummarizedExperiment class in R

.pull-left-80[&lt;img src="Figures/DataManagement/SummarizedExperiment.png" class="fancyimage size-60"&gt;]
.pull-right-30[**A visual overiview of a SummarizedExperiment. Notice the similarity to and *ExpressionSet* class. The *ExpressionSet* Class is the source of inspiration form Bioconductor *SummarizedExperimet* and *MultiAssayExperiment* owing to it's success.**]


---
## Demonstrate with real data


### Annotations from BioMart for Feature data

  - *txdb from biomart*
  - *Gene annotations from biomart*
 


### Summarized Experiment

 - *RowData from txdb object such as genomic coordinates and gene/transcript annotations*
 - *PhenoData such as patient IDs and characteristics*
 - *Assays such as counts, intensities, normalized counts*
 - *MetaData such as nature of experiment, how wa it performed*

### MultiAssayExperiment
 - *Create differnet experiments*

---
##  Multi Assay Experiment 
&lt;img src="Figures/DataManagement/MultiAssayExperiment.png" class="fancyimage size-150"&gt;
.pull-right-50[*ExperimentList*. *colData"*: **“primary”**, *sampleMap*, *metaData* ]


```r
mes&lt;- MultiAssayExperiment::MultiAssayExperiment()
slotNames(mes)
```

```
## [1] "ExperimentList" "colData"        "sampleMap"      "drops"         
## [5] "metadata"
```
---

## Create MultiAssayExperiment
 - *Create differnet experiments*
 
 - *ExperimentList*
 The datasets contained in elements of the ExperimentList can have:
  - column names (required)
  - row names (optional)
  
 - *colData"*: **“primary”**
  - metadata that describes the ‘biological unit’ which can refer to   
  specimens, experimental subjects, patients, etc. 
  - colData dataset must contain patient identifiers.
---

# Multi Assay Experiment (Cont)
 - *sampleMap* 
  - The sampleMap is a DataFrame that relates the “primary” data (colData) to the experimental 
  
  - assays:assay provides the names of the different experiments / assays performed. These are user-defined, with the only requirement that the names of the ExperimentList, where the experimental assays are stored, must be contained in this column.
  
  - primary provides the “primary” sample names. All values in this column must also be present in the rownames of - colData(MultiAssayExperiment). 
  
  - colname provides the sample names used by experimental datasets, which in practice are often different than the primary sample names. For each assay, all column names must be found in this column. 
  
 - *metadata*
  - Metadata can be added at different levels of the MultiAssayExperiment.
 Can be of ANY class, for storing study-wide metadata, such as citation information. 

---
## Save and loading the data (demonstration)

RMarkdown, Rdata, RDS objects to store and access data on disk

```r
fil &lt;- tempfile("women", fileext = ".rds")
#save a single object to file
#saveRDS(women, fil)
## restore it under a different name
#women2 &lt;- readRDS(fil)
#identical(women, women2)
## or examine the object via a connection, which will be opened as needed.
#con &lt;- gzfile(fil)
#readRDS(con)
#close(con)
```

---

---
## Thank You

**[Feedback form](https://forms.gle/DLFaw2c7DAx3gmDs9)**
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "Omics-Integration 2019 • %current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
